// stores/entity-store.js
var Fluxxor = require("fluxxor");
var actions = require("actions");

//utils
var sanitise  = require('utils/sanitise')

//helper
var _       = require('lodash')
var log     = require('debug')(':stores:entity-store')


var EntityStore = Fluxxor.createStore({
  initialize: function(options) {
    this.name = options.sheet
    this.type = options.type
    
    this.data = []    
    this.loading = false
    this.loadData = options.loadData
    //construct and call our load data closure
    
    this.loadData({
      key:options.key, 
      bucket:options.bucket, 
      sheet:options.sheet, 
      isProxy:options.isProxy
    }, this.loadDataSuccess)()
    
    this.loading = true
  },

  loadDataSuccess: function (data) {
    log('success', this.name)
    this.data = _.sortBy(sanitise(data[this.name].elements),function(d){
      return (this._isNormalInteger(d.id) ? parseInt(d.id) : d.id)
    },this)
    
//    _.each(this.data,function(d){
//      d.type = this.type
//    },this)
    
    this.loading = false
    this.emit('change')
  },

  isLoading : function(){
    return this.loading
  },

  findOne: function (match) {    
    return _.find(this.data, match)
  },
  query: function (match) {
    if (!match) return  this.data
    return _(this.data).where(match).clone(true)
  },
  filter: function (match) {
    if (!match) return  this.data
    return _(this.data).filter(match).clone(true)
  },
  entities: function () {
    return _.clone(this.data,true)
  },  
  countWhere: function(match) {
    return this.filter(match).length
  },
  _isNormalInteger : function (str) {
    var n = ~~Number(str);
    return String(n) === str && n >= 0;
  },
  getEntityType : function(){
    return this.type
  }

});

module.exports = EntityStore;
