// stores/entity-store.js
var Fluxxor = require("fluxxor");
var actions = require("actions");

//utils
var sanitise  = require('utils/sanitise')

//helper
var _       = require('lodash')
var log     = require('debug')(':stores:entity-store')


var EntityStore = Fluxxor.createStore({
  initialize: function(options) {
    this.name = options.sheet
    this.type = options.type
    
    this.data = []    
    this.loading = false
    this.loadData = options.loadData
    //construct and call our load data closure
    this.loadData(options.key, options.sheet, this.loadDataSuccess)()
    this.loading = true
  },

  loadDataSuccess: function (data) {
    log('success', this.name)
    log('success', data[this.name].elements)
    this.data = sanitise(data[this.name].elements)
    this._setType()
    this.loading = false
    this.emit('change')
  },
  _setType : function() {
    _.each(this.data,function(d){
      d.type = this.type
    },this)
  },
  isLoading : function(){
    return this.loading
  },

  findOne: function (match) {    
    return _.find(this.data, match)
  },
  query: function (match) {
    if (!match) return  this.data
    return _(this.data).where(match).clone(true)
  },
  filter: function (match) {
    if (!match) return  this.data
    return _(this.data).filter(match).clone(true)
  },
  entities: function () {
    return _.clone(this.data,true)
  },  
  countWhere: function(match) {
    return this.filter(match).length
  }
  

});

module.exports = EntityStore;
