// stores/print-store.js
// flux
var Fluxxor = require("fluxxor")
var actions = require("actions")
var createPdf = require('pdfmake-browserified')

// helpers
var _       = require('lodash')
var log     = require('debug')('src:node_modules:stores:print-store')

// Route Store binds route changing actions and triggers router transitions
var PrintStore = Fluxxor.createStore({
  initialize: function (options) {    
    
    // route actions
    this.bindActions(
      actions.constants.PRINT.ACTION, this.printAction
    )
  },
  printAction : function(payload){
    this.waitFor(["actions","issues","groups","agencies","pages"], 
    function(actionStore,issueStore,groupStore,agencyStore,pageStore) {
    
      var action = actionStore.findOne({ id: payload.id })    
      
      // get categories for entity 
      var issues = issueStore.filter(function(category){
        return action.issues.split(',').indexOf(category.id) > -1
      })
      
      
      var dd = {
        content: [
          { text: action.title, style: 'header1', margin: [0, 20] },
          { text: 'DESCRIPTION', style: 'header2', margin: [0, 10] },
          {
            columns: [
              {
                stack: [
                  { text: action.description },
                  { text: 'DESIRED OUTCOME', style: 'header2', margin: [0, 10] },
                  { text: action.desiredoutcome },
                  { text: 'MEASURES', style: 'header2', margin: [0, 10] },
                  { text: action.measures },
                  { text: 'TARGET DATE', style: 'header2', margin: [0, 10] },
                  { text: action.targetdate }
                ],
                width: '80%'
              },
              [
                { text: 'Issue', style: 'header3', margin: [0, 10] },
                { text: action.issues, style: 'blue' },
                { text: 'Population Groups', style: 'header3', margin: [0, 10] },
                { text: action.groups, style: 'blue' },
                { text: 'Government Agency', style: 'header3', margin: [0, 10] },
                { text: action.agencies, style: 'blue' }
              ]
            ]
          }
        ],

        styles: {
          header1: {
            fontSize: 22,
            bold: true
          },
          header2: {
            fontSize: 15,
            bold: true,
            color: '#7696A5'
          },
          header3: {
            fontSize: 12,
            bold: true
          },
          blue: {
            color: '#008399'
          }
        },
        defaultStyle: {
          columnGap: 20
        }
      }
      createPdf(dd).open()      
    });
    
  }



});

module.exports = PrintStore;
