// stores/print-store.js
// flux
var Fluxxor = require('fluxxor')
var actions = require('actions')
var createPdf = require('pdfmake-browserified')

// helpers
var _       = require('lodash')
var log     = require('debug')('src:node_modules:stores:print-store')

// Route Store binds route changing actions and triggers router transitions
var PrintStore = Fluxxor.createStore({
  initialize: function (options) {
    this.styles = {
      header1: {
        fontSize: 22,
        bold: true
      },
      header2: {
        fontSize: 15,
        bold: true,
        color: '#7696A5'
      },
      header3: {
        fontSize: 12,
        bold: true
      },
      blue: {
        color: '#008399'
      }
    }
    this.structure = {
      agency: {
        mainTitle: 'ABOUT',
        main: 'description',
        sideTitle: 'WEBSITE',
        side: 'link'
      },
      group: {
        mainTitle: 'COMMENTARY',
        main: 'commentary',
        sideTitle: 'ABOUT THIS GROUP',
        side: 'description'
      },
      issue: {
        mainTitle: 'CURRENT CONTEXT',
        main: 'currentcontext',
        sideTitle: 'OTHER ISSUES RAISED BY CIVIL SOCIETY',
        side: 'otherissuesraised'
      },
    /*  treatyBody: {
        mainTitle: 'ABOUT',
        main: 'description',
        sideTitle: 'CONCLUDING OBSERVATIONS',
        side: 'concludingobservations'
      }*/
    }
    // route actions
    this.bindActions(
      actions.constants.PRINT, this.print
    )
  },

  print: function (payload) {
    this.waitFor(['actions', 'issues', 'groups', 'agencies', 'pages'],
    function (actionStore, issueStore, groupStore, agencyStore, pageStore) {
      var content, category, relatedEntities
      switch (payload.type) {
        case 'action' :
          var action = actionStore.findOne({ id: payload.id })
          content = this.getAction(action)
          break
        case 'issue' :
          category = issueStore.findOne({id: payload.id})
          relatedEntities = issueStore.query({issues: category.id})
          content = this.getCategory(payload, category, relatedEntities)
          break
        case 'group' :
          category = groupStore.findOne({id: payload.id})
          relatedEntities = actionStore.query({groups: category.id})
          content = this.getCategory(payload, category, relatedEntities)
          break
        case 'agency' :
          category = agencyStore.findOne({id: payload.id})
          relatedEntities = actionStore.query({agencies: category.id})
          content = this.getCategory(payload, category, relatedEntities)
          break
      }
      createPdf(content).download()
    })
  },

  getAction: function (action) {
    // get categories for entity
    //var issues = issueStore.filter(function (category) {
    //  return action.issues.split(',').indexOf(category.id) > -1
    //})
    var dd = {
      content: [
        {text: action.title, style: 'header1', margin: [0, 20]},
        {text: 'DESCRIPTION', style: 'header2', margin: [0, 10]},
        {
          columns: [
            {
              stack: [
                {text: action.description},
                {text: 'DESIRED OUTCOME', style: 'header2', margin: [0, 10]},
                {text: action.desiredoutcome},
                {text: 'MEASURES', style: 'header2', margin: [0, 10]},
                {text: action.measures},
                {text: 'TARGET DATE', style: 'header2', margin: [0, 10]},
                {text: action.targetdate}
              ],
              width: '80%'
            },
            [
              {text: 'Issue', style: 'header3', margin: [0, 10]},
              {text: action.issues, style: 'blue'},
              {text: 'Population Groups', style: 'header3', margin: [0, 10]},
              {text: action.groups, style: 'blue'},
              {text: 'Government Agency', style: 'header3', margin: [0, 10]},
              {text: action.agencies, style: 'blue'}
            ]
          ]
        }
      ],
      styles: this.styles,
      defaultStyle: {
        columnGap: 20
      }
    }
    return dd
  },

  getCategory: function (payload, category, relatedEntities) {
    var relatedEntitiesIndex = _.map(relatedEntities, function (value, key) {
      return {columns: [{width: '10%', text: value.id}, {width: '90%', text: value.title, style: 'index'}]}
    })

    var relatedEntitiesAction = _.map(relatedEntities, function (value, key) {
      return {stack: this.getAction(value).content, pageBreak: 'before'}
    }, this)

    var dd = {
      content: [
        {text: category.title, style: 'header1', margin: [0, 20], pageBreak: 'after'},
        {text: category.title, style: 'header1', margin: [0, 20]},
        {
          columns: [
            {
              stack: [
                {text: this.structure[payload.type].mainTitle, style: 'header2', margin: [0, 10]},
                {text: category[this.structure[payload.type].main]}
              ],
              width: '80%'
            },
            [
              {text: this.structure[payload.type].sideTitle, style: 'header3', margin: [0, 10]},
              {text: category[this.structure[payload.type].side], style: 'blue'}
            ]
          ],
          pageBreak: 'after'
        },
        {text: 'ACTION INDEX', style: 'header1', margin: [0, 20]},
        {text: relatedEntities.length + ' RELATED ACTIONS', style: 'header2', margin: [0, 20]},
        {stack: relatedEntitiesIndex},
        {stack: relatedEntitiesAction}
      ],
      styles: this.styles,
      defaultStyle: {
        columnGap: 20
      }
    }
    return dd
  }
})

module.exports = PrintStore
