// components/views/categories/categories.js
// react
var React           = require('react')
var Fluxxor         = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

//components
var EntityList = require('components/list/EntityList')
var Load       = require('components/load')
var NavPrimary = require('components/nav/nav-categories')

//util
var referencing = require('utils/referencing')

// bootstrap
var Grid  = require('react-bootstrap/lib/Grid')
var Row 	= require('react-bootstrap/lib/Row')
var Col 	= require('react-bootstrap/lib/Col')

//helpers
var _   = require('lodash')
var log = require('debug')('src:components:explore-by')

// display actions and recommendations by "category"
module.exports = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React),
    StoreWatchMixin('issues', 'groups', 'agencies','treatybodies','recommendations','actions')
  ],

  getStateFromFlux: function () {
    var flux = this.getFlux()   
    
    // store loading conditions
    return {
      issuesLoading:      
           flux.store('issues').isLoading() 
        || flux.store("actions").isLoading()
        || flux.store("recommendations").isLoading(),
      groupsLoading:      
           flux.store('groups').isLoading()
        || flux.store("actions").isLoading()
        || flux.store("recommendations").isLoading(),
      agenciesLoading:    
           flux.store('agencies').isLoading()
        || flux.store("actions").isLoading(),
      treatybodiesLoading:
           flux.store('treatybodies').isLoading()        
        || flux.store("recommendations").isLoading()
    }
  },

  
  render: function () {
    log ('loading state ', this.state)
    var category = this.props.params.id
    // show actions and recommendations by issues
    if (category === 'issues' && !this.state.issuesLoading) {
      // render both actions AND recommendation columns
      return this.renderContent(
        ['actions','recommendations'],
        'countActions'
      )              
    } 
    else if (category === 'groups' && !this.state.groupsLoading) {        
      // render both actions AND recommendation columns
      return this.renderContent(
        ['actions','recommendations'],
        'countActions'
      )
    }
    else if (category === 'agencies' && !this.state.agenciesLoading) {         
      // render both actions AND recommendation columns
      return this.renderContent(
        ['actions'],        
        'countActions'
      )      
      
    }    
    else if (category === 'treatybodies' && !this.state.treatybodiesLoading) {
      return this.renderContent(
        ['recommendations'],
        'countRecs'
      )         
    } else { 
      return (
        <div>
          <NavPrimary 
            active = {category}
          />
          <Load/>
        </div>
      )
    } 
      
                 
    
  },
  // sortBy : 'countActions'  
  renderContent : function (types,sortBy){        
    var flux = this.getFlux() 
    
    var category = this.props.params.id
    var title = flux.store(category).getTitle()
        
    // get all category items and prep with entity-counts
    // sort by number of actions/recs (reverse order)    
    var entities = _.sortBy(this.getEntities(category,types),sortBy).reverse()
    // maximum value of any related entity (actions or recommendation)
    var maxValue = this.getMaxCount(entities,types)
    var maxValueActions = this.getMaxCount(entities,['actions'])
    var maxValueRecs = this.getMaxCount(entities,['recommendations'])
    // the table columns
    var cols = this.getColumns(title,types,maxValue)    
    var colsXS = this.getColumnsXS(title,types,maxValueActions,maxValueRecs)    

    
    return (
      <div>
      <NavPrimary 
        active = {category}
      />
      <div className="page-content-explore page-content">
        <Grid className='page-section explore-section-table'>
              <h1>{'Explore by ' + title}</h1>
              <EntityList
                collapsible = {false}
                className="explore-table hidden-xs"
                cols = {cols}
                entities = {entities}
                entityType = {flux.store(category).getEntityType()}                         
                link = {true}
              />
              <EntityList
                collapsible = {false}
                className="explore-table visible-xs"
                cols = {colsXS}
                entities = {entities}
                entityType = {flux.store(category).getEntityType()}                         
                link = {true}
              />
              <div className="table-note">
                <span className="text-uppercase strong">Note: </span>
                Actions and recommendations may not add up to totals due to multiple possible assignments
              </div>
        </Grid>
      </div>
      </div>
    )
  },
  // get category items
  getEntities : function (category,types) {
    var flux = this.getFlux() 
    var maxValue = 0
    return _.map(flux.store(category).entities(),function(entity){                    
          
      if (types.indexOf('actions') > -1) {  
        //get actions for issue, issue referenced by action
        entity.countActions = referencing(flux,{
          entity : entity,
          attr : category,
          from : 'actions'
        }).length     
      } 
      if (types.indexOf('recommendations') > -1) {  
        entity.countRecs = referencing(flux,{
          entity : entity,
          attr : category,
          from : 'recommendations'
        }).length          

        entity.countRecsAccepted = _.filter(referencing(flux,{
          entity : entity,
          attr : category,
          from : 'recommendations'
        }),function(item){
          return item.response === 'Accepted'
        }).length  
      }
      return entity      
    })
  },
  
  // figure out maximum number of entities for given category
  getMaxCount : function(entities,types){
    var maxVal = 0    
    _.each(entities,function(entity){
      if (types.indexOf('actions') > -1) {  
        maxVal = Math.max(maxVal,entity.countActions) 
      }
      if (types.indexOf('recommendations') > -1) {  
        maxVal = Math.max(maxVal,entity.countRecs) 
      }      
    })    
    return maxVal
    
  },
  
  // set up entity table columns
  getColumns : function(title, types, maxValue){
        
    var col_title = {
      type:   'text',
      title:  title,
      attr:   'title',
//      nestAfter : {
//        visibleXS : true,
//        cols:[
//          _.extend({},col_actions,{hiddenXS:false}),
//          _.extend({},col_recs,{hiddenXS:false})
//        ]
//      }
    }  
    
    // action column if actions, else spacer
    var col_actions = (types.indexOf('actions') > -1) 
      ? {
          type:       'bar',
          title:      'Government Actions',
          value :     {attr:'countActions',color:'#0095AC'}, 
          maxValue :  maxValue,
        }
      : {
          type:'spacer-bar'
        }
    
    // rec column if recs, else spacer
    var col_recs = (types.indexOf('recommendations') > -1) 
      ? {
          type:       'ratio-bar',
          title:      'UPR Recommendations',
          ratio :     {attr:'countRecsAccepted',color:'#E19900'}, 
          value :     {attr:'countRecs',color:'#AEBCC1'}, 
          maxValue :  maxValue,
        }
      : {
          type:'spacer-bar'
        }
    // key column if recs, else spacer
    var col_links = 
        (types.indexOf('recommendations') > -1 
      && types.indexOf('actions') === -1) 
        ? { 
            type: 'key-links',
            key: {
              type:'units',
              items: [{
                  title:'Accepted',
                  color:'#E19900'
                },
                {
                  title:'Not Accepted',
                  color:'#AEBCC1'
                }
              ] 
            }             
          }
        : { 
            type: 'links',
          }     
      
    // key column if recs, else spacer
    var col_links = (types.indexOf('recommendations') > -1) 
      ? { 
          type: 'key-links',
          key: {
            type:'units',
            items: [{
                title:'Accepted',
                color:'#E19900'
              },
              {
                title:'Not Accepted',
                color:'#AEBCC1'
              }
            ] 
          }             
        }
      : { 
          type: 'links'
        }             
    
    return [col_title, col_actions, col_recs, col_links]
  },
  // set up entity table columns
  getColumnsXS : function(title, types, maxValueActions, maxValueRecs){
        
    var col_title = {
      type:   'text',
      title:  title,
      attr:   'title'
    }  
       
    var col_bar = (types.indexOf('actions') > -1)                
    ? {
        type:       'bar',
        title:      'Actions',
        value :     {attr:'countActions',color:'#0095AC'}, 
        maxValue :  maxValueActions,
      }
    : {
        type:'spacer-bar',
        hiddenXS : true
      }   
    // key column if recs and not actions, else spacer
    var col_links = { 
      type: 'links',
    }           
    
                  
    return [col_title, col_bar, col_links]
  }
  
})
