// components/views/pages/intro/intro.js
//react
var React           = require('react')
var Fluxxor         = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

//components
var Load            = require('components/load')
var SelectTaxonomy  = require('components/nav/select-taxonomy')

//bootstrap
var Grid 				= require('react-bootstrap/lib/Grid')
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')

//helpers
var _   = require('lodash')
var log = require('debug')('src:components:intro')


module.exports = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React),
    StoreWatchMixin('actions','recommendations')    
  ],
  getStateFromFlux: function () {
    var flux = this.getFlux()    
    return {
      loading : flux.store("recommendations").isLoading()
             || flux.store("actions").isLoading()
    }
  },
  render: function () {

    log('props, state', this.props, this.state)
    var flux = this.getFlux()    

    var navItems = [
      {slug:'issues',title:'Issue',active:false},
      {slug:'groups',title:'Population Group',active:false},
      {slug:'agencies',title:'Government Agency',active:false},
      {slug:'treatybodies',title:'Treaty Body',active:false}
    ]

    
    return (
      <div role='main' className='intro text-center'>
        <Grid>
          <Row role="article" className="intro-content">
            <Col sm={10} smPush={1}>
              <h1 className='serif'>
                <div role="presentation" >New Zealand Human Rights</div>
                <div role="presentation" >National Plan of Action 2015 &ndash; 2019</div>
              </h1>
              <p>
                This National Plan sets out the actions the Government is taking 
                as a result of the commitments it made through its second Universal Periodic Review (UPR)
                before the United Nations Human Rights Council in 2014
              </p>
            </Col>
          </Row>
        </Grid>
        <Grid>
          <Row className="intro-stats hidden-xs" 
              aria-live="polite" 
              aria-busy={this.state.loading ? "true" : "false"}>
            {this.renderStats()}
          </Row>
        </Grid>
        <Grid className="intro-explore-hint">                      
          <h4 role="presentation" className="serif intro-explore">Explore actions and recommendations by</h4>
        </Grid>
        <div className="intro-issue-select nav-primary">  
          <Grid>         
            <SelectTaxonomy 
              flux = {flux }
              items = {navItems}           
              type = 'exploreby'
            />
          </Grid>
        </div>
        <Grid key={2} >
          <div className="intro-read-more">
            <p><a href="#" onClick={this.handleNavClick('about')}>or learn more about the UPR process</a></p>
          </div>
        </Grid>
      </div>
    )
  },

  renderStats : function (){
    if (this.state.loading) {
      return  (<Load alt="Loading statistics &hellip;"/>)
    } else { 
      var flux = this.getFlux()    
      var no_submissions = 54
      var no_recs     = flux.store('recommendations').entities().length
      var no_actions  = flux.store('actions').entities().length
      
      return(            
        <Col sm={10} smPush={1}>              
          <ul role="presentation" className="list-inline">
            <li role="presentation" className="submission-stat intro-stat-item">
              <div role="presentation" className="stat-large">{no_submissions}</div>
              <div role="presentation" className="stat-title">Submissions to the UPR</div>
              <div role="presentation" className="stat-content">by 250+ civil society individuals and groups</div>
            </li>
            <li role="presentation" className="intro-stat-divider">
              <span className="icon-arrow-tall"/>
            </li>
            <li role="presentation" className="recs-stat intro-stat-item">
              <div role="presentation" className="stat-large">{no_recs}</div>
              <div role="presentation" className="stat-title">UPR Recommendations</div>
              <div role="presentation" className="stat-content">of which 121 were accepted by the Government</div>
            </li>
            <li role="presentation" className="intro-stat-divider">
              <span className="icon-arrow-tall"/>
            </li>            
            <li role="presentation" className="action-stat intro-stat-item">
              <div role="presentation" className="stat-large">{no_actions}</div>
              <div role="presentation" className="stat-title">Government Actions</div>
              <div role="presentation" className="stat-content">address the UPR recommendations</div>
            </li>
          </ul>
        </Col> 
      )
    }
  },
  handleNavClick: function (page) {
    var flux = this.getFlux() 

    return function (e) {
      e.preventDefault()
      flux.actions.navPage(page)
    }
  },          
})
