// components/views/pages/intro/intro.js
//react
var React           = require('react')
var Fluxxor         = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

//components
var Load            = require('components/load')
var SelectTaxonomy  = require('components/nav/select-taxonomy')

//bootstrap
var Grid 				= require('react-bootstrap/lib/Grid')
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')

//helpers
var _   = require('lodash')
var log = require('debug')('src:components:intro')


module.exports = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React),
    StoreWatchMixin('actions','recommendations')
  ],
  getStateFromFlux: function () {
    var flux = this.getFlux()
    return {
      loading : flux.store("recommendations").isLoading()
             || flux.store("actions").isLoading()
    }
  },
  render: function () {

    log('props, state', this.props, this.state)
    var flux = this.getFlux()

    var navItems = [
      {slug:'issues',title:'Issue',active:false},
      {slug:'groups',title:'Population Group',active:false},
      {slug:'agencies',title:'Government Agency',active:false},
      {slug:'treatybodies',title:'Treaty Body',active:false}
    ]


    return (
      <div role='main' className='intro text-center'>
        <div className='background-blue'>
          <Grid>
            <Row role="article" className="intro-content">
              <Col sm={10} smPush={1}>
                <p>
                  This National Plan sets out the actions the Government is taking
                  as a result of the commitments it made through its second Universal Periodic Review (UPR)
                  before the United Nations Human Rights Council in 2014
                </p>
              </Col>
            </Row>
          </Grid>
          <Grid>
            <Row className="intro-stats hidden-xs">
              {this.renderStats()}
            </Row>
          </Grid>
        </div>
        <span className="arrow"/>
        <div className='background-light-blue'>
          <Grid className="intro-explore-hint">
            <h4 className="intro-explore">Explore by category</h4>
            <Row className="intro-explore-text">
              <Col sm={10} smPush={1} md={6} mdPush={3}>
                <p>Actions and recommendations have been categorised along four dimensions to provide further insight and analysis</p>
              </Col>
            </Row>
          </Grid>
          <div className="intro-issue-select nav-primary">
            <Grid>
              <SelectTaxonomy
                flux = {flux }
                items = {navItems}
                type = 'exploreby'
              />
            </Grid>
          </div>
        </div>
        <div className=''>
          <Grid className="intro-explore-all">
            <h4 className="intro-explore">Explore all actions and recommendations</h4>
            <Row className="intro-explore-text">
              <Col sm={10} smPush={1} md={10} mdPush={1}>
                <p>Complete lists of actions and recommendations as well as advanced filter and print options</p>
              </Col>
            </Row>
            <Row className="intro-explore-btn">
              <Col sm={10} smPush={1} md={3} mdOffset={2}>
                <p><a href="#" className="btn" onClick={this.handleNavClick('all')}>All Actions</a></p>
              </Col>
              <Col sm={10} smPush={1} md={3}>
                <p><a href="#" className="btn" onClick={this.handleNavClick('recommendation')}>All recommendations</a></p>
              </Col>
            </Row>
          </Grid>
        </div>
      </div>
    )
  },

  renderStats : function (){
    if (this.state.loading) {
      return  (<Load />)
    } else {
      var flux = this.getFlux()
      var no_submissions = 54
      var no_recs     = flux.store('recommendations').entities().length
      var no_actions  = flux.store('actions').entities().length

      return(
        <Col sm={10} smPush={1}>
          <ul className="list-inline">
            <li className="submission-stat intro-stat-item">
              <div className="stat-large">{no_submissions}</div>
              <div className="stat-title">Submissions to the UPR</div>
              <div className="stat-content">by 250+ civil society individuals and groups</div>
            </li>
            <li className="intro-stat-divider">
              <span className="icon-arrow-tall"/>
            </li>
            <li className="recs-stat intro-stat-item">
              <div className="stat-large">{no_recs}</div>
              <div className="stat-title">UPR Recommendations</div>
              <div className="stat-content">of which 121 were accepted by the Government</div>
            </li>
            <li className="intro-stat-divider">
              <span className="icon-arrow-tall"/>
            </li>
            <li className="action-stat intro-stat-item">
              <div className="stat-large">{no_actions}</div>
              <div className="stat-title">Government Actions</div>
              <div className="stat-content">address the UPR recommendations</div>
            </li>
          </ul>
        </Col>
      )
    }
  },
  handleNavClick: function (page) {
    var flux = this.getFlux()

    return function (e) {
      e.preventDefault()
      flux.actions.navPage(page)
    }
  },
})
