// components/views/intro/Intro.js
//react
var React           = require('react')
var Fluxxor         = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

//components
var Icon        = require('components/Icon')
var Load            = require('components/Load')
var SelectTaxonomy  = require('components/nav/SelectTaxonomy')

//bootstrap
var Grid 				= require('react-bootstrap/lib/Grid')
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')

//helpers
var _   = require('lodash')
var log = require('debug')('src:components:intro')


module.exports = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React),
    StoreWatchMixin('actions','recommendations')
  ],
  getStateFromFlux: function () {
    var flux = this.getFlux()
    return {
      loading : flux.store("recommendations").isLoading()
             || flux.store("actions").isLoading()
    }
  },
  render: function () {

    log('props, state', this.props, this.state)
    var flux = this.getFlux()

    var navItems = [
      {slug:'issues',title:'Issues',active:false},
      {slug:'groups',title:'Population Groups',active:false},
      {slug:'agencies',title:'Government Agencies',active:false},
      {slug:'treatybodies',title:'UN Treaty Bodies',active:false}
    ]


    return (
      <div role='main' className='intro text-center'>
        <section className="intro-lead">
          <Grid>
            <Row role="article" className="intro-content">
              <Col sm={10} smPush={1} md={8} mdPush={2}>
                <p>
                  This National Plan sets out the actions the Government is taking 
                  to protect and promote Human Rights as a result of the commitments 
                  it made through its second Universal Periodic Review (UPR) before 
                  the United Nations Human Rights Council in 2014                  
                </p>
              </Col>
            </Row>
            <Row className="intro-stats hidden-xs"
              aria-live="polite"
              aria-busy={this.state.loading ? "true" : "false"}>
              {this.renderStats()}
            </Row>
            <span className="arrow"/>
          </Grid>     
        </section>
        <section className="intro-category-explore">
          <Grid>
            <h3 className="intro-explore">Explore by Category</h3>
            <Row className="intro-explore-text">
              <Col sm={10} smPush={1} md={8} mdPush={2}>
                <p>
                  Government Actions and UPR Recommendations have been 
                  categorised across four groupings to provide further 
                  insight and analysis
                </p>
              </Col>
            </Row>
            <SelectTaxonomy
              flux = {flux }
              items = {navItems}
              type = 'categories'
            />
          </Grid>
        </section>
        <section className="intro-explore-all">  
          <Grid>
            <h3 className="intro-explore">Actions and Recommendations</h3>
            <Row className="intro-explore-text">
              <Col sm={10} smPush={1} md={8} mdPush={2}>
                <p>Complete lists of Government Actions and UPR Recommendations as well as custom filter and print options</p>
              </Col>
            </Row>
            <Row className="intro-explore-buttons">
              <Col xs={6} >
                <a href="#" className="btn pull-right" onClick={this.handleItemClick({slug:'',route:'actions'})}>
                  <span className="btn-inner">Government Actions</span>
                </a>
              </Col>
              <Col xs={6} >
                <a href="#" className="btn pull-left" onClick={this.handleItemClick({slug:'',route:'recommendations'})}>
                  <span className="btn-inner">UPR Recommendations</span>
                </a>
              </Col>
            </Row>
          </Grid>
        </section>
      </div>
    )
  },

  renderStats : function (){
    if (this.state.loading) {
      return  (<Load />)
    } else {
      var flux = this.getFlux()
      var no_submissions = 54
      var no_recs     = flux.store('recommendations').entities().length
      var no_actions  = flux.store('actions').entities().length

      return(
        <Col md={10} mdPush={1}>
          <ul role="presentation" className="list-inline">
            <li role="presentation" className="submission-stat intro-stat-item">
              <div role="presentation" className="stat-large">{no_submissions}</div>
              <div role="presentation" className="stat-title">Submissions to the UPR</div>
              <div role="presentation" className="stat-content">by 250+ civil society individuals and groups</div>
            </li>
            <li role="presentation" className="intro-stat-divider">
              <Icon name="arrow-large-r"/>
            </li>
            <li role="presentation" className="recs-stat intro-stat-item">
              <div role="presentation" className="stat-large">{no_recs}</div>
              <div role="presentation" className="stat-title">UPR Recommendations</div>
              <div role="presentation" className="stat-content">of which 121 were accepted by the Government</div>
            </li>
            <li role="presentation" className="intro-stat-divider">
              <Icon name="arrow-large-r"/>
            </li>
            <li role="presentation" className="action-stat intro-stat-item">
              <div role="presentation" className="stat-large">{no_actions}</div>
              <div role="presentation" className="stat-title">Government Actions</div>
              <div role="presentation" className="stat-content">to address UPR Recommendations</div>
            </li>
          </ul>
        </Col>
      )
    }
  },
  handleItemClick: function (item) {
    return _.bind(function (e) {
      e.preventDefault()
      this.getFlux().actions.navigate(item.route,item.slug)
    },this)
  }
})
