// components/list/CategoryList.js
var React       = require('react')
var Fluxxor     = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin




//helpers
var _           = require('lodash')
var log         = require('debug')('src:components:entityList')



var CategoryList = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React)
  ],
  propTypes : {
    entities: React.PropTypes.array,
    entityType : React.PropTypes.string
  },
  
  render: function () {    
    log('props',this.props)
    
    return (
      <ul className={'hrc-category-list'}>
        {
          _.map(this.props.entities,function(item, i){
            return (
              <li key={i}>
                <a href="#" onClick={this.handleCategoryClick(item.id,this.props.entityType)}>
                {item.title}
                </a>
              </li>
            )
          },this)
        }
      </ul>
    )
      
  },
  handleCategoryClick: function (id,type) {

    return function (e) {
      e.preventDefault()
      log('handleCategoryClick',type + '/' + id )
      this.getFlux().actions.entity.select(type,id)
    }.bind(this)
  },    
})

module.exports = CategoryList;
