// components/list/EntityList.js
//main
var React     = require('react')

//components
var TableBody            = require('components/list/TableBody')
var TableHead            = require('components/list/TableHead')

// bootstrap
var Table 				= require('react-bootstrap/lib/Table')


//helpers
var _           = require('lodash')
var log         = require('debug')('src:components:entityList')



var EntityList = React.createClass({

  propTypes : {
    cols: React.PropTypes.array,    
    entities: React.PropTypes.array,    
    className: React.PropTypes.string,
    linked:React.PropTypes.bool,
    collapsible:React.PropTypes.bool,
    maxItems:React.PropTypes.number,
    showMoreLabel: React.PropTypes.string,
    entityType : React.PropTypes.string
  },

  getInitialState: function () {
    return {
      collapsed: true
    }
  },
  componentWillReceiveProps: function (nextProps) {
    //collapse main text when switching between indicators
    if (!this.state.collapsed && nextProps.id !== this.props.id) {
      this.setState({ collapsed: true })
    }
  },  
  render: function () {    
    log('props',this.props)
    
    var className = typeof this.props.className !== 'undefined' ? this.props.className : ''
    var collapsible = typeof this.props.collapsible !== 'undefined' ? this.props.collapsible : true
    var maxItems = typeof this.props.maxItems !== 'undefined' ? this.props.maxItems : 5

      
    var entities = (collapsible & this.state.collapsed) 
    ? this.props.entities.slice(0,maxItems) 
    : this.props.entities
    
    var ShowMore = null
    
    if (this.props.entities.length > maxItems && collapsible) {
      var ShowMore = (
        <div className="show-more">
          <a href="#" onClick={this.toggleCollapse}>
            <div className="show-more-note">
              {this.state.collapsed 
                  ? "Showing " + maxItems + " of " + this.props.entities.length + " " + this.props.showMoreLabel
                  : "Showing all " + this.props.entities.length + " " + this.props.showMoreLabel
                }
            </div>  
            <div className="show-more-hint strong">
              {this.state.collapsed ? 'Show all' : 'Show less'}
            </div>
          </a>
        </div>
      )
    }
    
    return (
     <div>
      <Table className={"hrc-table " + className }>
        <TableHead 
          cols = {this.props.cols} />
        <TableBody 
          linked = {this.props.linked}
          cols = {this.props.cols}
          entities = {entities}
          entityType = {this.props.entityType} />           
      </Table>
      { ShowMore}
      </div>
    )
      
  },
  
  toggleCollapse: function (e) {
    e.preventDefault()
    this.setState({ collapsed: !this.state.collapsed })
  }
        
  
})

module.exports = EntityList;
