// components/list/EntityListFactory.js
var React       = require('react')
var Fluxxor     = require('fluxxor')

// components
var EntityList = require('components/list/EntityList')

//helpers
var _           = require('lodash')
var log         = require('debug')('src:components:EntityListFactory')



var EntityListFactory = React.createClass({
  
  mixins: [
    Fluxxor.FluxMixin(React)  
  ],
  
  propTypes : {
    relatedType: React.PropTypes.string,
    relatedEntities : React.PropTypes.array,
    showEmpty : React.PropTypes.bool,
    showCount: React.PropTypes.bool,
    collapsible : React.PropTypes.bool, 
    linked : React.PropTypes.bool
  },
  
  render: function () {    
    log('props',this.props)
    
    var relatedEntities  = this.props.relatedEntities
    var relatedType = this.props.relatedType
    var store = this.getFlux().store(relatedType)
    
    var cols = this.getColumns(relatedType)
    
    if (relatedEntities.length === 0){
      if (!this.props.showEmpty) {
        return null       
      } else {
        return (   
          <h2>{
            (this.props.showCount 
              ? relatedEntities.length + ' ' 
              : '') 
            + 'Related ' + store.getTitle(relatedEntities.length)}
          </h2>
        )
      }
    } else {        
      return (<div>
          <h2>
            {(this.props.showCount 
              ? relatedEntities.length + ' ' 
              : '') 
            + 'Related ' + store.getTitle(relatedEntities.length)}
          </h2>
          <EntityList
            showMoreLabel={relatedType}
            collapsible={this.props.collapsible}  
            className={relatedType + '-table'}
            linked = {this.props.linked}
            cols = {cols}            
            entities = {relatedEntities} 
            entityType = {relatedType}
          />
      </div>)
    }
      
  },
  
  getArticleColumns : function () {
    return [
        {type:'no', title: 'No.', attr:'id'},            
        {type:'text', title: 'UDHR Article',  attr:'title'}
      ]
  },
  
  getActionColumns : function () {
    return [
        {type:'no',   title: 'Id',      attr:'id'},                        
        {type:'text', title: 'Action', attr:'title'},
        {type:'links'}
      ]
  },
  
  getRecommendationColumns : function () {
    var recs = this.props.relatedEntities
    
    var countRecsAccepted = recs.filter(function(entity) {
      return entity.response === 'Accepted'
    }).length
    
    var countRecsRejected = recs.length - countRecsAccepted        
    
    return [
            {type:'no', title: 'No.', attr:'id'},            
            {type:'text', title: 'UPR Recommendation', attr:'title'},
            {type:'icon', title: 'Response', 
             attr:'response', class: function(response){
                // function will be evaluated in cell
                return (response === 'Accepted') ? 'icon-accepted' : 'icon-rejected'  
              }
            },
            { 
              type: 'key-links',               
              key: {
                type:'no',
                items: [{
                    title:'Accepted',
                    no: countRecsAccepted,
                    color:'#E19900'
                  },
                  {
                    title:'Not Accepted',
                    no: countRecsRejected,
                    color:'#AEBCC1' 
                  }
                ] 
              }, hiddenXS:true             
            },
            {type:'links',visibleXS:true}
          ]
  },
  
  getColumns : function(relatedType) {

    // articles columns
    if (relatedType === 'articles'){
      return this.getArticleColumns()
    } 
    
    // actions columns
    else if(relatedType === 'actions') {
      return this.getActionColumns()      
    }
    
    // recommendation columns
    else if(relatedType === 'recommendations') {
      return this.getRecommendationColumns()      
    } 
            
  }
})

module.exports = EntityListFactory;
