// components/list/EntityTable.js
//main
var React     = require('react')

//components
var TableBody            = require('components/list/TableBody')
var TableHead            = require('components/list/TableHead')

// bootstrap
var Table 				= require('react-bootstrap/lib/Table')


//helpers
var _           = require('lodash')
var log         = require('debug')('src:components:entityList')



var EntityTable = React.createClass({

  propTypes : {
    cols: React.PropTypes.array,    
    entities: React.PropTypes.array,    
    className: React.PropTypes.string,
    linked:React.PropTypes.boolean,
    collapsible:React.PropTypes.boolean,
    maxItems:React.PropTypes.number
  },

  getInitialState: function () {
    return {
      collapsed: true
    }
  },
  componentWillReceiveProps: function (nextProps) {
    //collapse main text when switching between indicators
    if (!this.state.collapsed && nextProps.id !== this.props.id) {
      this.setState({ collapsed: true })
    }
  },  
  render: function () {    
    log('props',this.props)
    
    var className = typeof this.props.className !== 'undefined' ? this.props.className : ''
    var collapsible = typeof this.props.collapsible !== 'undefined' ? this.props.collapsible : true
    var maxItems = typeof this.props.maxItems !== 'undefined' ? this.props.maxItems : 10
    
    var ShowMore = (
      <div className="show-more">
        <a href="#" onClick={this.toggleCollapse}>{this.state.collapsed ? 'Show more' : 'Show less'}</a>
      </div>
    )
      
    var entities = (collapsible & this.state.collapsed) 
    ? this.props.entities.slice(0,maxItems) 
    : this.props.entities
      
    return (
            <div>
      <Table className={"hrc-table " + className }>
        <TableHead 
          cols = {this.props.cols} />
        <TableBody 
          linked = {this.props.linked}
          cols = {this.props.cols}
          entities = {entities} />
      </Table>
      { (this.props.entities.length > maxItems && collapsible) ? ShowMore : null }
      </div>
    )
      
  },
  
  toggleCollapse: function (e) {
    e.preventDefault()
    this.setState({ collapsed: !this.state.collapsed })
  }
        
  
})

module.exports = EntityTable;
