// components/list/EntityList.js
//main
var React     = require('react')

//components
var TableBody            = require('components/list/TableBody')
var TableHead            = require('components/list/TableHead')

var ShowMore = require('components/show-more')
//helpers
var _           = require('lodash')
var log         = require('debug')('src:components:entityList')



var EntityList = React.createClass({

  propTypes : {
    cols: React.PropTypes.array,    
    entities: React.PropTypes.array,    
    className: React.PropTypes.string,
    linked:React.PropTypes.bool,
    collapsible:React.PropTypes.bool,
    maxItems:React.PropTypes.number,
    showMoreLabel: React.PropTypes.string,
  },

  getInitialState: function () {
    return {
      collapsed: true
    }
  },
  componentWillReceiveProps: function (nextProps) {
    //collapse main text when switching between indicators
    if (!this.state.collapsed && nextProps.id !== this.props.id) {
      this.setState({ collapsed: true })
    }
  },  
  render: function () {    
    log('props',this.props)
    
    var className = typeof this.props.className !== 'undefined' ? this.props.className : ''
    var collapsible = typeof this.props.collapsible !== 'undefined' ? this.props.collapsible : true
    var maxItems = typeof this.props.maxItems !== 'undefined' ? this.props.maxItems : 5

      
    var visibleEntities = (collapsible & this.state.collapsed) 
    ? this.props.entities.slice(0,maxItems) 
    : this.props.entities
    
    return (
      <div className="hrc-table-wrap">
        <table className={"hrc-table " + className }>
          <caption>{this.props.caption}</caption>
          <TableHead 
            cols = {this.props.cols} />
          <TableBody 
            linked = {this.props.linked}
            cols = {this.props.cols}
            rows = {visibleEntities}
          />
        </table>
        <ShowMore
          toggleCollapse = {this.toggleCollapse}    
          collapsed = {this.state.collapsed}
          collapsible = {collapsible}
          maxItems = {maxItems}
          totalItems = {this.props.entities.length}
          showMoreLabel = {this.props.showMoreLabel}
        />
      </div>
    )
      
  },
  
  toggleCollapse: function (e) {
    e.preventDefault()
    this.setState({ collapsed: !this.state.collapsed })
  }
        
  
})

module.exports = EntityList;
