// components/list/TableLinkRow.js
//react
var React   = require('react')
var Fluxxor = require('fluxxor')

// components
var TableCellIcon		  = require('components/list/TableCellIcon')
var TableCellBar		  = require('components/list/TableCellBar')
var TableCellRatioBar	= require('components/list/TableCellRatioBar')

//helpers
var _      = require('lodash')
var marked = require('utils/parse-markdown')
var log    = require('debug')('src:components:entityListRow')



var TableLinkRow = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React)
  ],
  propTypes : {
    entity: React.PropTypes.object,    
    entityType: React.PropTypes.string,    
    cols: React.PropTypes.array,
    linked:React.PropTypes.bool    
  },
  
  render: function () {    
    log('props',this.props)
    
    var entity = this.props.entity
    
    var linked = typeof this.props.linked !== 'undefined' ? this.props.linked : true

    var className = (linked) ? 'hrc-tr hrc-table-row hrc-table-link-row': 'hrc-tr hrc-table-row'
    return (
            
        <div className={className} 
             onClick={(linked) ? this.handleRowClick(entity.id,this.props.entityType) : false} >
          {
            _.map(this.props.cols,function(col,i){
              return (
                <div key={i} className={'hrc-td hrc-cell hrc-cell-'+col.type}>
                  {this.renderCell(entity,col)}
                </div>
              )
            },this)
          }
        </div>
           
    )
  },
  renderCell:function(entity,col) {
    if ( col.type === 'text') {
      return ( <div className="marked-inner-html"
              dangerouslySetInnerHTML={{
                __html: marked(entity[col.attr], {sanitize: true})
              }} >
            </div>
      )
    } else if ( col.type === 'no' || col.type === 'label') {
      return entity[col.attr]    
    } else if ( col.type === 'icon') {
      return (
              <TableCellIcon
                col={col}
                entity={entity} />                
            )
    } else if ( col.type === 'bar' ) {      
      return (
              <TableCellBar
                col={col}
                entity={entity} />                
            )
      
    } else if ( col.type === 'ratio-bar' ) {
      
      return (
              <TableCellRatioBar
                col={col}
                entity={entity} />                
            ) 
      
    } else if ( col.type === 'links' 
            || col.type === 'key-links'  ) {
      return (
        <span className="icon-arrow-link"></span>
      )
    } else {
      return null
    }
  },
  
  
  handleRowClick: function (id,type) {

    return function (e) {
      e.preventDefault()
      log('handleRowclick',type + '/' + id )
      this.getFlux().actions.entity.select(type,id)
    }.bind(this)
  },  
})


module.exports = TableLinkRow;
