// components/list/TableLinkRow.js
//main
var React     = require('react')
var React       = require('react')
var Fluxxor     = require('fluxxor')

// bootstrap


//helpers
var _           = require('lodash')
var log         = require('debug')('src:components:entityList')



var TableLinkRow = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React)
  ],
  propTypes : {
    entity: React.PropTypes.object,    
    cols: React.PropTypes.array    
  },
  
  render: function () {    
    log('props',this.props)
    
    var entity = this.props.entity
    return (
            
        <tr className={'hrc-table-link-row'} onClick={this.handleRowClick(entity.id,entity.type)} >
          {
            _.map(this.props.cols,function(col){
              return (
                <td>{entity[col]}</td>
              )
            })
          }
        </tr>
           
    )
  },
  handleRowClick: function (id,type) {

    return function (e) {
      e.preventDefault()
      log('handlerowclick',type + '/' + id )
      this.getFlux().actions.entity.select(this.props.entity.type,id)
    }.bind(this)
  },  
})


module.exports = TableLinkRow;
