// components/subjects/action/action.js
var React           = require('react')
var Fluxxor         = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

//components

// bootstrap
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')

//helpers

var _               = require('lodash')

var log       = require('debug')('src:components:subjects:action')

module.exports = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React),
    StoreWatchMixin('actions','recommendations','issues','groups','agencies')
  ],

  getStateFromFlux: function () {
    var flux = this.props.flux
    return {
      entitiesLoading : flux.store("actions").isLoading(),
      recsLoading : flux.store("recommendations").isLoading(),
      issuesLoading: flux.store('issues').isLoading(),
      groupsLoading : flux.store("groups").isLoading(),
      agenciesLoading : flux.store("agencies").isLoading()      
    }
  },

  render: function () {
    var flux = this.props.flux

    return (
      <Row>
        { this.renderMain() }
      </Row>
    )
  },
  renderMain : function (){
    var flux = this.props.flux
    var entity = flux.stores.actions.findOne({ id: this.props.params.id })
log('action',entity)
    return (
      <Col md={12}>
        <p>Action </p>
        <p>
          { (this.state.entitiesLoading) ? 'loading' : entity.title }
        </p>
      </Col>
    )
  },

})
