// components/subjects/action/action.js
var React       = require('react')
var Fluxxor     = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

//components
var Load        = require('components/load')
var EntityTable            = require('components/list/EntityTable')
var CategoryList            = require('components/list/CategoryList')

// bootstrap
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')

//helpers
var _           = require('lodash')
var marked          = require('marked')

var log         = require('debug')('src:components:explore:issue')

module.exports = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React),
    StoreWatchMixin('issues', 'actions','recommendations','articles')
  ],

  getStateFromFlux: function () {
    var flux = this.props.flux
    return {
      entitiesLoading: flux.store('actions').isLoading(),     
      issuesLoading : flux.store("issues").isLoading(),
      agenciesLoading : flux.store("agencies").isLoading(),
      groupsLoading : flux.store("groups").isLoading(),
      recsLoading : flux.store("recommendations").isLoading()
      
      
    }
  },

   render: function () {
    var flux = this.props.flux
    
    if (this.state.entitiesLoading 
    || this.state.issuesLoading
    || this.state.groupsLoading
    || this.state.recsLoading 
    || this.state.agenciesLoading 
    ) {
      return  (
        <Load />
      )
    } else {
      
       var entity = flux.stores.actions.findOne({ id: this.props.params.id })

      return (
        <div>
           {this.renderMain(entity)}
           {this.renderRecs(entity)}
        </div>        
      )
    }
  },
  renderMain : function (entity){
    var flux = this.props.flux
    return (
      [
        <Row>
          <Col md={12}>
            <h2>Action</h2>
            <h1>{entity.title}</h1>
          </Col>
        </Row>,
        <Row>
          <Col md={8}>
            <h4>Desciption</h4>
            <div
              dangerouslySetInnerHTML={{
                __html: marked(entity.description, {sanitize: true})
              }} >
            </div>
            <h4>Desired Outcome</h4>
            <div
              dangerouslySetInnerHTML={{
                __html: marked(entity.desiredoutcome, {sanitize: true})
              }} >
            </div>
            <h4>Measures</h4>
            <div
              dangerouslySetInnerHTML={{
                __html: marked(entity.measures, {sanitize: true})
              }} >
            </div>
            <h4>Target date</h4>
            <div
              dangerouslySetInnerHTML={{
                __html: marked(entity.targetdate, {sanitize: true})
              }} >
            </div>  
          </Col>
          <Col md={4}>  
            {this.renderIssues(entity)}
            {this.renderGroups(entity)}
            {this.renderAgencies(entity)}
            
          </Col>
        </Row>
      ]      
    )
  },
  
  renderRecs : function (entity){
    var flux = this.props.flux
    
    var recs =  flux.store('recommendations').filter(function(item){
      return entity.recommendations.split(',').indexOf(item.id) > -1
    })    
    
    return ([
      <h2>Related UPR Recommendations</h2>,
      <EntityTable
        cols = {[
          {title: 'ID',      field:'id'},            
          {title: 'UPR Recommendations', field:'title'}
        ]}          
        entities = {recs} />            
      
    ])
  },
  renderIssues : function (entity){
    var flux = this.props.flux
    
    var items =  flux.store('issues').filter(function(item){
      return entity.issues.split(',').indexOf(item.id) > -1
    })
    return ([
      <h5>Issues</h5>,
      <CategoryList
        entities = {items} />
    ])
  },
  renderGroups : function (entity){
    var flux = this.props.flux
    
    var items =  flux.store('groups').filter(function(item){
      return entity.groups.split(',').indexOf(item.id) > -1
    })
    return ([
      <h5>Population Groups</h5>,
      <CategoryList
        entities = {items} />
    ])
  },
  renderAgencies : function (entity){
    var flux = this.props.flux
    
    var items =  flux.store('agencies').filter(function(item){
      return entity.agencies.split(',').indexOf(item.id) > -1
    })
    return ([
      <h5>Responsible Agency</h5>,
      <CategoryList
        entities = {items} />
    ])
  },
 

})
