// components/views/singles/action/action.js
// react
var React           = require('react')
var Fluxxor         = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

//components
var Load                = require('components/load')
var EntityListFactory   = require('components/list/EntityListFactory')
var CategoryListFactory = require('components/list/CategoryListFactory')
var NavPrimary          = require('components/nav/nav-categories')
var NavBack             = require('components/nav/nav-back')

//util
var referencing = require('utils/referencing')

// bootstrap
var Grid 				= require('react-bootstrap/lib/Grid')
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')

//helpers
var _           = require('lodash')
var marked      = require('utils/parse-markdown')
var log         = require('debug')('src:components:action')

// display single action
module.exports = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React),
    StoreWatchMixin('issues', 'actions','recommendations','agencies','groups')
  ],

  getStateFromFlux: function () {
    var flux = this.getFlux() 
    return {
      loading: flux.store('actions').isLoading()
            || flux.store("issues").isLoading()
            || flux.store("agencies").isLoading()
            || flux.store("groups").isLoading()
            || flux.store("recommendations").isLoading()
    }
  },

  render: function () {
    var flux = this.getFlux() 
    
    if (this.state.loading) {
      return (
        <div>
          <NavPrimary />
          <Load/>
        </div>
      )
    } else {
      
      // get action from store
      var entity = flux.stores.actions.findOne({ id: this.props.params.id })
      
      // get recommendations for action     
      // entity references recommendations
      var items = referencing(flux, {
        entity : entity,
        attr : 'recommendations',
        to : 'recommendations'
      }) 
      
      // - action content      
      // - related recommendations
      return (
        <div>
          <NavPrimary />              
          <div className="page-content-action page-content">        
            {this.renderMain(entity)}
            <Grid className={'page-section page-section-recommendations'}>
              <EntityListFactory
                relatedType = {'recommendations'}
                relatedEntities = {items}
                showEmpty = {true}
                showCount = {true}
                collapsible = {true} 
                linked = {true}
              />
            </Grid>
          </div>
      </div>
      )
    }
  },
  renderMain : function (entity){
    var flux = this.getFlux() 
    return (
      <Grid className='page-section action-section-analysis'>
        <Row key={0}  className="header-row">
          <Col sm={10}>
            <h2 className="header-type-large">{'Action '} &ndash; {entity.id}</h2>
            <h3 className="header-title">{entity.title}</h3>
          </Col>
          <div className='nav-control'>
            <NavBack/>    
          </div> 
        </Row>
        <Row key={1}  className="content-main">
          <Col sm={8} className="column-smart">
            {this.renderSMART(entity)}
            <h5>Description</h5>
            <div className="marked-inner-html"
              dangerouslySetInnerHTML={{
                __html: marked(entity.description, {sanitize: true})
              }} >
            </div>
            <h5>Desired Outcome</h5>
            <div className="marked-inner-html"
              dangerouslySetInnerHTML={{
                __html: marked(entity.desiredoutcome, {sanitize: true})
              }} >
            </div>
            <h5>Measures</h5>
            <div className="marked-inner-html custom-list-icon"
              dangerouslySetInnerHTML={{
                __html: marked(entity.measures, {sanitize: true})
              }} >
            </div>
            <h5>Target date</h5>
            <div className="marked-inner-html strong"
              dangerouslySetInnerHTML={{
                __html: marked(entity.targetdate, {sanitize: true})
              }} >
            </div>  
          </Col>
          <Col sm={4} className="column-categories">  
            <CategoryListFactory 
              entity = {entity}
              store = {flux.store('issues')}
              taxonomy = {'issues'}
            /> 
            <CategoryListFactory 
              entity = {entity}
              store = {flux.store('groups')}
              taxonomy = {'groups'}
            />
            <CategoryListFactory 
              entity = {entity}
              store = {flux.store('agencies')}
              taxonomy = {'agencies'}
            />    
          </Col>
        </Row>
      </Grid>
    )
  },
  checkCriterion : function(field) {
    return field === 1 || field === '1' || field === 'y' || field === 'yes'
  },
  renderSMART : function(entity) {
    
    var criteria = [
      {id:'s',label:'Specific',active:this.checkCriterion(entity.isspecific)},
      {id:'m',label:'Measurable',active:this.checkCriterion(entity.ismeasurable)},
      {id:'a',label:'Assignable',active:this.checkCriterion(entity.isassignable)},
      {id:'r',label:'Result-oriented',active:this.checkCriterion(entity.isresultoriented)},
      {id:'t',label:'Time-bound',active:this.checkCriterion(entity.istimebound)}
    ]
    
    return ([
      <h5  key={0} className='tt-none'>SMART criteria: Assessment coming soon</h5>,
      <a  key={1} href="#" title="About S.M.A.R.T. criteria" onClick={this.handleNavClick('smart')} className='smart-information'>
        <span className="icon-information"/>
      </a>,
      <ul key={2} className="list-inline list-bordered" style={{width:'100%'}}>      
        {
          _.map(criteria,function(item, i){
            return this.renderSMARTcriterion(item)
          },this)
        }
      </ul>
    ])
    
  }, 
  
  renderSMARTcriterion : function (criterion) {
    var className_li = "list-item-smart " + (criterion.active ? "active" : "")
    var className_icon = "icon-smart-" + criterion.id
    return ([
        <li className={className_li} style={{width:'20%'}}>
          <div className={className_icon}></div>
          <div className="smart-label hidden-xs">{criterion.label}</div>
        </li>
    ])
  },
  
  
  handleNavClick: function (page) {
    var flux = this.getFlux()

    return function (e) {
      e.preventDefault()
      flux.actions.navPage(page)
    }
  },
})
