// components/table/EntityTable.js
//main
var React     = require('react')

//components
var TableBody            = require('components/table/TableBody')
var TableHead            = require('components/table/TableHead')

//helpers
var _           = require('lodash')
var log         = require('debug')('src:components:entityList')



var EntityTable = React.createClass({

  propTypes : {
    cols: React.PropTypes.array,    
    rows: React.PropTypes.array,    
    className: React.PropTypes.string,
    linked:React.PropTypes.bool,
    caption:React.PropTypes.string,
    handleRowClick:React.PropTypes.func
  },

  componentDidMount: function() {
    if (typeof this.props.caption !== 'undefined' && this.props.caption !== '' ) {
      this.getDOMNode().setAttribute('summary', "Table: " + this.props.caption)
    }      
  },
  render: function () {    
//    log('props',this.props)
    
    var className = typeof this.props.className !== 'undefined' ? this.props.className : ''
    var caption = typeof this.props.caption !== 'undefined' && this.props.caption !== '' 
      ? (<caption><h2>{this.props.caption}</h2></caption>)
      : null
    return (
      <table className={"hrc-table " + className } >
        {caption}
        <TableHead 
          cols = {this.props.cols} />
        <TableBody 
          linked = {this.props.linked}
          cols = {this.props.cols}
          rows = {this.props.rows}          
          handleRowClick = {this.props.handleRowClick}    
        />
      </table>
    )      
  },
        
  
})

module.exports = EntityTable;
