// components/explore/group/group.js
var React       = require('react')
var Fluxxor     = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

//components
var Load        = require('components/load')
var EntityTable            = require('components/list/EntityTable')

// bootstrap
var Grid 				= require('react-bootstrap/lib/Grid')
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')

//helpers
var _           = require('lodash')
var marked               = require('utils/parse-markdown')

var log         = require('debug')('src:components:explore:issue')

module.exports = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React),
    StoreWatchMixin('groups', 'actions','recommendations')
  ],

  getStateFromFlux: function () {
    var flux = this.props.flux
    return {
      entitiesLoading: flux.store('groups').isLoading(),
      actionsLoading : flux.store("actions").isLoading(),
      recsLoading : flux.store("recommendations").isLoading()
      
    }
  },

   render: function () {
    var flux = this.props.flux
    
    if (this.state.entitiesLoading 
    || this.state.actionsLoading
    || this.state.recsLoading 
    ) {
      return  (
        <Load />
      )
    } else {
      
       var entity = flux.stores.groups.findOne({ id: this.props.params.id })

      return (
      <div className="page-content-group page-content">
        <Grid className='page-section group-section-analysis'>
           {this.renderMain(entity)}
        </Grid>
        <Grid className='page-section group-section-actions'>   
           {this.renderActions(entity)}
        </Grid>
        <Grid className='page-section group-section-recs'>   
           {this.renderRecs(entity)}
        </Grid>
      </div>     
      )
    }
  },
  renderMain : function (entity){
    var flux = this.props.flux
    return (
      [
        <Row className="header-row">
          <Col md={12}>
            <h5 className="header-type">Group</h5>
            <h1 className="header-title">{entity.title}</h1>
          </Col>
        </Row>,
        <Row className="content-main">
          <Col md={8}>
            <h5>Commentary</h5>
            <div className="marked-inner-html"
              dangerouslySetInnerHTML={{
                __html: marked(entity.commentary, {sanitize: true})
              }} >
            </div>
          </Col>
          <Col md={4}>
           { entity.description.trim() !== '' 
            ? (
              [<h5>About this Group</h5>,
              <div className="marked-inner-html small custom-list-icon"
                dangerouslySetInnerHTML={{
                  __html: marked(entity.description, {sanitize: true})
                }} >
              </div>] )
             : null
           }
          </Col>
        </Row>
      ]      
    )
  },
  renderActions : function (entity){
    var flux = this.props.flux
    
    var actions =  flux.store('actions').filter(function(item){
              return item.groups.split(',').indexOf(entity.id) > -1
            })
    if (actions.lenth === 0){
      return null       
    } else {    
      return ([
        <h2>{actions.length + ((actions.length === 1) ? ' Action' : ' Actions')}</h2>,
        <EntityTable
          className="action-table"   
          collapsible={false}
          cols = {[
            {type:'no',   title: 'No',      attr:'id'},                        
            {type:'text', title: 'Action', attr:'title'},
            {type:'links'}
          ]}     
          entities = {actions} />   
      ])
    }
  },
  renderRecs : function (entity){
    var flux = this.props.flux
    
    var recs =  flux.store('recommendations').filter(function(item){
              return item.groups.split(',').indexOf(entity.id) > -1
            })
    if (recs.length === 0 ) {
      return null
    } else {            
      return ([
        <h2>{recs.length + ((recs.length === 1) ? ' Related UPR Recommendation' : ' Related UPR Recommendations')}</h2>,
        <EntityTable
          showMoreLabel="related UPR recommendations"
          className="upr-recommendations-table"
          cols = {[
            {type:'no', title: 'No',      attr:'id'},            
            {type:'text', title: 'UPR Recommendation', attr:'title'},
            {type:'icon', title: 'Response', class: function(entity){
                return (entity.response === 'Accepted') ? 'icon-accepted' : 'icon-rejected'  
              }
            },
            { 
                type: 'key-links',               
                key: {
                  type:'icons',
                  items: [{
                      title:'Accepted',
                      class:'icon-accepted-sm'
                    },
                    {
                      title:'Not Accepted',
                      class:'icon-rejected-sm'
                    }
                  ] 
                }             
              }          
          ]}  
          entities = {recs} />           

      ])
    }
  }
  
 

})
