// components/explore/issue/issue.js
var React           = require('react')
var Fluxxor         = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

//components

// bootstrap
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')

//helpers

var _               = require('lodash')

var log       = require('debug')('src:components:explore:issue')

module.exports = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React),
    StoreWatchMixin('issues', 'actions','recommendations','articles')
  ],

  getStateFromFlux: function () {
    var flux = this.props.flux
    return {
      entitiesLoading: flux.store('issues').isLoading(),
      actionsLoading : flux.store("actions").isLoading(),
      recsLoading : flux.store("recommendations").isLoading(),
      articlesLoading : flux.store("articles").isLoading(),
      
    }
  },

   render: function () {
    var flux = this.props.flux

    return (
      <Row>
        { this.renderMain() }
      </Row>
    )
  },
  renderMain : function (){
    var flux = this.props.flux
    var entity = flux.stores.issues.findOne({ id: this.props.params.id })
    log('issue',entity)
    return (
      <Col md={12}>
        <p>Issue </p>
        <p>
          { (this.state.entitiesLoading) ? 'loading' : entity.title }
        </p>
      </Col>
    )
  },
 

})
