// components/explore/issue/issue.js
var React       = require('react')
var Fluxxor     = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

//components
var Load        = require('components/load')
var EntityTable            = require('components/list/EntityTable')

// bootstrap
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')

//helpers
var _           = require('lodash')
var marked          = require('marked')

var log         = require('debug')('src:components:explore:issue')

module.exports = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React),
    StoreWatchMixin('issues', 'actions','recommendations','articles')
  ],

  getStateFromFlux: function () {
    var flux = this.props.flux
    return {
      entitiesLoading: flux.store('issues').isLoading(),
      actionsLoading : flux.store("actions").isLoading(),
      recsLoading : flux.store("recommendations").isLoading(),
      articlesLoading : flux.store("articles").isLoading(),
      
    }
  },

   render: function () {
    var flux = this.props.flux
    
    if (this.state.entitiesLoading 
    || this.state.actionsLoading
    || this.state.recsLoading 
    || this.state.articlesLoading 
    ) {
      return  (
        <Load />
      )
    } else {
      
       var entity = flux.stores.issues.findOne({ id: this.props.params.id })

      return (
        <div>
           {this.renderMain(entity)}
           {this.renderActions(entity)}
           {this.renderRecs(entity)}
           {this.renderArticles(entity)}
        </div>        
      )
    }
  },
  renderMain : function (entity){
    var flux = this.props.flux
    return (
      [
        <Row>
          <Col md={12}>
            <p>Issues &gt;</p>
            <h1>{entity.title}</h1>
          </Col>
        </Row>,
        <Row>
          <Col md={8}>
            <h4>Current Context</h4>
            <div
              dangerouslySetInnerHTML={{
                __html: marked(entity.currentcontext, {sanitize: true})
              }} >
            </div>
          </Col>
          <Col md={4}>
            <h4>Other Issues Raised by Civil Society</h4>
            <div
              dangerouslySetInnerHTML={{
                __html: marked(entity.otherissuesraised, {sanitize: true})
              }} >
            </div>  
          </Col>
        </Row>
      ]      
    )
  },
  renderActions : function (entity){
    var flux = this.props.flux
    
    var actions =  flux.store('actions').query({issue:entity.id})
    
    return ([
      <h2>Actions</h2>,
      <EntityTable
        cols = {['id','title']}            
            entities = {actions} />      
    ])
  },
  renderRecs : function (entity){
    var flux = this.props.flux
    
    var recs =  flux.store('recommendations').query({issue:entity.id})
    return ([
      <h2>Related UPR Recommendations</h2>,
      <EntityTable
        cols = {['id','title']}
            entities = {recs} />            
      
    ])
  },
  renderArticles : function (entity){
    var flux = this.props.flux
    
    var articles =  flux.store('articles').filter(function(item){
      return entity.articles.split(',').indexOf(item.id) > -1
    })
    return ([
      <h2>Related Articles</h2>,
      <EntityTable
        cols = {['id','title']}
            entities = {articles} />         
    ])
  },
 

})
