// printPdf/layouts/action-layout.js
// helpers
var _ = require('lodash')

// util
var parseHTML = require('../parse-html')

module.exports = function (data) {
  
  function getContentItem(item){
    switch (item.type) {
      case 'html' :
        return parseHTML(item.content)
        break
      case 'text' :
        return item.content
        break
      case 'targetDate' :
        return _.map(item.content,function(c,i){
          return i === 0 ? { text : c, style : ['bold','paragraph']} : c
        })
        break          
    }    
  }      
  function makeCategoryList (categories, split) {
    split = typeof split !== 'undefined' ? split : ' | '
    return _.map(categories, function (cat) { return cat }).join(split)
  }  
  var elements = {    
    header : function(id) {
      return {
        columns : [
          {
            stack: [ { text: 'GOVERNMENT ACTION', style: 'itemSupTitle' } ],
            width:427
          },
          {
            table : {
              widths: [ '100%'],
              body : [
                [{ text: 'ACTION', style : 'pageIdTableCellType' }],
                [ { text: id,  style : 'pageIdTableCellId' }]
              ]
            },
            layout: 'noBorders',
            style : 'pageIdTable',
            width:63
          }
        ]
      }
    },
    content : function(contentList) {
      var content = []   
      _.each(contentList,function(item){
        if (!_.isEmpty(item.content)) {
          content.push({text: item.title.toUpperCase(), style: ['itemContentSectionTitle','green'] })
          content.push(getContentItem(item))
        }

      })
      return content
    },
    smart : function(smart) {      
      return {
        table : {
          widths: [ 113,50],
          body : [
            [{ text: 'SMART criteria', colSpan: 2, style : ['sideTableCellTitle','sideTableCell'] },''],
            [ {canvas: [{ type: 'line', x1: 0, y1: 0, x2: 150, y2: 0, lineWidth: 0.5, lineColor :'#fff' }],
                colSpan:2,style:'sideTableCellLineSecondary'},''],
            [ 
              {
                text :'Specific',
                style : ['sideTableCellTitleSecondary']   
              },
              {
                image: smart.isspecific ? 'smart-s-pass-active-bg' : 'smart-s-fail-active-bg', 
                rowSpan : 2,width: 25,style : ['sideTableCellTitleSecondary']   
              }
            ],
            [ 
              {
                text : smart.isspecific ? 'Met' : 'Not met',
                style : ['sideTableCell', smart.isspecific ? 'green' : 'grey' ]   
              },
              ''
            ],
            [ {canvas: [{ type: 'line', x1: 0, y1: 0, x2: 150, y2: 0, lineWidth: 0.5, lineColor :'#fff' }],
                colSpan:2,style:'sideTableCellLineSecondary'},''],
            [ 
              {
                text :'Measurable',
                style : ['sideTableCellTitleSecondary']   
              },
              {
                image: smart.ismeasurable ? 'smart-m-pass-active-bg' : 'smart-m-fail-active-bg', 
                rowSpan : 2,width: 25,style : ['sideTableCellTitleSecondary']      
              }
            ],
            [ 
              {
                text : smart.ismeasurable ? 'Met' : 'Not met',
                style : ['sideTableCell', smart.ismeasurable ? 'green' : 'grey' ]   
              },
              ''
            ],
            [ {canvas: [{ type: 'line', x1: 0, y1: 0, x2: 150, y2: 0, lineWidth: 0.5, lineColor :'#fff' }],
                colSpan:2,style:'sideTableCellLineSecondary'},''],
            [ 
              {
                text :'Assignable',
                style : ['sideTableCellTitleSecondary']   
              },
              {
                image: smart.isassignable ? 'smart-a-pass-active-bg' : 'smart-a-fail-active-bg', 
                rowSpan : 2,width: 25,style : ['sideTableCellTitleSecondary']     
              }
            ],
            [ 
              {
                text : smart.isassignable ? 'Met' : 'Not met',
                style : ['sideTableCell', smart.isassignable ? 'green' : 'grey' ]   
              },
              ''
            ],
            [ {canvas: [{ type: 'line', x1: 0, y1: 0, x2: 150, y2: 0, lineWidth: 0.5, lineColor :'#fff' }],
                colSpan:2,style:'sideTableCellLineSecondary'},''],
            [ 
              {
                text :'Results-oriented',
                style : ['sideTableCellTitleSecondary']   
              },
              {
                image: smart.isresultoriented ? 'smart-r-pass-active-bg' : 'smart-r-fail-active-bg', 
                rowSpan : 2,width: 25,style : ['sideTableCellTitleSecondary']      
              }
            ],
            [ 
              {
                text : smart.isresultoriented ? 'Met' : 'Not met',
                style : ['sideTableCell', smart.isresultoriented ? 'green' : 'grey' ]   
              },
              ''
            ],
            [ {canvas: [{ type: 'line', x1: 0, y1: 0, x2: 150, y2: 0, lineWidth: 0.5, lineColor :'#fff' }],
                colSpan:2,style:'sideTableCellLineSecondary'},''],
            [ 
              {
                text :'Timebound',
                style : ['sideTableCellTitleSecondary']   
              },
              {
                image: smart.istimebound ? 'smart-t-pass-active-bg' : 'smart-t-fail-active-bg', 
                rowSpan : 2,width: 25,style : ['sideTableCellTitleSecondary']      
              }
            ],
            [ 
              {
                text : smart.istimebound ? 'Met' : 'Not met',
                style : ['sideTableCell', smart.istimebound ? 'green' : 'grey' ]   
              },
              ''
            ]
          ]
        },
        layout: 'noBorders',
        style : 'sideTable',
        width:170
      }
    },
    categories : function(categories){
      return{
        table : {
          widths: [ 170],
          body : [
            [{ text: 'Action Categories', style : ['sideTableCellTitle','sideTableCell'] }],
            [ { canvas: [{ type: 'line', x1: 0, y1: 0, x2: 150, y2: 0, lineWidth: 0.5, lineColor :'#fff' }],
                style:'sideTableCellLine'}],
            [ { text : categories.issues.length === 1 ? 'Issue' : 'Issues', style : ['sideTableCellTitleSecondary'] }],
            [ { text : makeCategoryList(categories.issues), style : ['sideTableCell', 'green' ] } ],
            [ { canvas: [{ type: 'line', x1: 0, y1: 0, x2: 150, y2: 0, lineWidth: 0.5, lineColor :'#fff' }],
                style:'sideTableCellLineSecondary'}],            
            [ { text : categories.groups.length === 1 ? 'Population Group' : 'Population Group', style : ['sideTableCellTitleSecondary'] }],
            [ { text : makeCategoryList(categories.groups), style : ['sideTableCell', 'green' ] }],
            [ { canvas: [{ type: 'line', x1: 0, y1: 0, x2: 150, y2: 0, lineWidth: 0.5, lineColor :'#fff' }],
                style:'sideTableCellLineSecondary'}],            
            [ { text : categories.agencies.length === 1 ? 'Government Agency' : 'Government Agencies', style : ['sideTableCellTitleSecondary'] }],
            [ { text : makeCategoryList(categories.agencies), style : ['sideTableCell', 'green' ] } ]
          ]
        },
        layout: 'noBorders',
        style : 'sideTable',
        width:170
      }        
      
    },
    recommendations : function(recs){
      return {
        table : {
          widths: [ 170],
          body : [
            [ { text: 'Addressed UPR Recommendation' + (recs.length !== 1 ? 's' : ''), style : ['sideTableCellTitle','sideTableCell'] }],
            [ {canvas: [{ type: 'line', x1: 0, y1: 0, x2: 150, y2: 0, lineWidth: 0.5, lineColor :'#fff' }],style:'sideTableCell'}],
            [ { text : 'No.', style : ['sideTableCellTitleSecondary'] } ],
            [ { text : makeCategoryList(recs,', '), style : ['sideTableCell', 'green' ] }]
          ]
        },
        layout: 'noBorders',
        style : 'sideTable',
        width :170
      }
    }
  }
  
  return [
    elements.header(data.id),
    {
      text: data.title,
      style: 'itemTitleAction'
    },
    {
      columns: [
        {
          stack: [
            {
              canvas: [{ type: 'line', x1: 0, y1:0, x2: 320, y2: 0, lineWidth: 0.5, lineColor :'#B5BEC0' }],
              margin: [0, 0, 0, 10]
            },
            elements.content(data.content)                      
          ],
          width: 320
        },
        {
          stack : [
            elements.smart(data.smart),
            elements.categories(data.categories),
            elements.recommendations(data.recommendations)
          ]          
        }                  
      ]
    }      
  ]  
}
