// printPdf/layouts/category-layout.js
// helpers
var _ = require('lodash')

// util
var parseHTML = require('../parse-html')

module.exports = function (data, type, showRecs) {
     
  function makeCategoryList (categories, split) {
    split = typeof split !== 'undefined' ? split : ' , '
    return _.map(categories, function (cat) { return cat }).join(split)
  }  
  var elements = {    
    header : function(data,type) {
      var idIcon 
      switch (type) {
        case 'issue' :
          idIcon = 'cat-issues-white'
          break
        case 'group' :
          idIcon = 'cat-groups-white'
          break
        case 'agency' :
          idIcon = 'cat-agencies-white'
          break
      }
      return {
        columns : [
          {
            stack: [ { text: data.supTitle.toUpperCase(), style: 'itemSupTitle' } ],
            width:427
          },
          {
            image : idIcon,
            width:63
          }
        ]
      }
    },
    aside : function(data) {
      if (data.aside.trim() === '') {
        return ''
      } else {
        return {
          table : {
            widths: [ 170],
            body : [
              [ { text: data.asideTitle , style : ['sideTableCellTitle','sideTableCell','sideTableCellCategory'] }],
              [ {canvas: [{ type: 'line', x1: 0, y1: 0, x2: 150, y2: 0, lineWidth: 0.5, lineColor :'#fff' }],
                  style:['sideTableCellLine','sideTableCellCategory']}],
              [ 
                {
                  stack : [
                    type !== 'agency' ? parseHTML(data.aside) : data.aside
                  ],
                  style : ['sideTableCell','sideTableCellCategory']
                }
              ]
            ]
          },
          layout: 'noBorders',
          style : 'sideTable',
          width :170
        }      
      }      
    },
    recommendations : function(showRecs, recs){
      if (showRecs) {
        return {
          table : {
            widths: [170],
            body : [
              [ { text: 'Related UPR Recommendation' + (recs.length !== 1 ? 's' : ''), 
                  style : ['sideTableCellTitle','sideTableCell','sideTableCellCategory'] }],
              [ {canvas: [{ type: 'line', x1: 0, y1: 0, x2: 150, y2: 0, lineWidth: 0.5, lineColor :'#fff' }],
                  style:['sideTableCellLine','sideTableCellCategory']}],
              [ { text : 'No.', 
                  style : ['sideTableCellTitleSecondary','sideTableCellCategory'] } ],
              [ { text : makeCategoryList(recs,', '), 
                  style : ['sideTableCell', 'blue','sideTableCellCategory' ] }]
            ]
          },
          layout: 'noBorders',
          style : 'sideTable',
          width :170
        }
      } else {
        return ''
      }
    }
  }
  
  showRecs = showRecs && typeof data.recommendationsList !== 'undefined' && data.recommendationsList.length > 0
  
  return [
    elements.header(data,type),
    {
      text: data.title,
      style: 'itemTitleCategory'
    },
    {
      columns: [
        {
          stack: [
            {
              canvas: [{ type: 'line', x1: 0, y1:0, x2: 320, y2: 0, lineWidth: 0.5, lineColor :'#B5BEC0' }],
              margin: [0, 0, 0, 10]
            },
            {text: data.contentTitle, style: ['itemContentSectionTitle','blue'] },           
            parseHTML(data.content)     
          ],
          width: 320
        },
        {
          stack : [
            elements.aside(data),            
            elements.recommendations(showRecs, data.recommendationsList)
          ]          
        }                  
      ]
    }      
  ]  
}
