// printPdf/layouts/title-layout.js
// helpers
var _ = require('lodash')

module.exports = function (data, type, sections) {
  
  function getFilters(filterData,type) {
    if (type === 'custom') {
      return _.map(filterData,function(filter){
        return {
          stack : [
            {text:filter.title.toUpperCase(),style:['titleFilterListTitle'],margin:[0,8,0,0]},
            {stack:_.map(filter.rows,function(row){
                return {text:row,style:['titleFilterListItem']}
              })
            }
          ],
          style:['title']
        }
      })
    } else {
      return []
    }
  }
  
  return [
    {
      columns : [
        {
          stack : [
            {text: 'New Zealand\'s National Plan of Action', style: ['titleHeaderTitle'],margin: [0, 0, 0,1]},
            {text: 'Mahere Rautāki a Motu', style: ['titleHeaderSubtitle']},
          ],
          style:['title','titleHeader'],
          width : 415
        },
        {
          text : data.reportType.toUpperCase(),
          alignment:'right',
          width : 100,
          style : ['title','titleReportType']
        }
      ],
      margin: [0, 0, 0, 10]
    },    
    {canvas: [{type: 'line', x1: 0, y1: 0, x2: 200, y2: 0, lineWidth: 0.5, lineColor: '#ffffff'}], margin: [0,0,0,100]},
    {
      columns: [
        {
          stack: [
            {text: data.supTitle.toUpperCase(), style: ['title','titleReportSupTitle'],margin: [0, 0, 0, 5]},
            {text: data.reportTitle, style: ['title','titleReportTitle'],margin:[0,0,0,10]}            
          ].concat(getFilters(data.filterData,type)),
          width: 514
        },        
        {
          canvas: [{type: 'line', x1: 0, y1: 0, x2:0, y2: 320, lineWidth: 0, lineColor: '#029BE5'}],
          width : 1
        }        
      ]
    },
    {canvas: [{type: 'line', x1: 0, y1: 0, x2:200, y2: 0, lineWidth: 0.5, lineColor: '#ffffff'}], margin: [0, 0, 0, 10]},
    {
      columns: [
        {
          stack: [
            {text: 'IN THIS REPORT', style: ['title', 'titleInReport'], margin: [0, 0, 0, 10]},
            {
              stack : _.map(sections,function(section){
                return {
                  stack : [
                    {text : section.title, style:'bold'},
                    {
                      stack : _.map(section.content, function(item){
                        return {text:item}
                      })
                    }
                  ],
                  margin : [0,0,0,10]
                }
              }),
              style :['title']
            }
          ]
        },        
        {
          canvas: [{type: 'line', x1: 0, y1: 0, x2:0, y2: 162, lineWidth:0, lineColor: '#029BE5'}],
          width : 1
        } 
      ]
    },
    {
      columns: [
        {
          stack: [
            {canvas: [{type: 'line', x1: 0, y1: 0, x2:200, y2: 0, lineWidth: 0.5, lineColor: '#ffffff'}], margin: [0, 0, 0, 15]},
            {text: [{text: 'DATE: ', style: ['bold']}, {text: data.date}], margin: [0, 0, 0, 5]},            
            {text: [{text: 'ISBN: ', style: ['bold']}, {text: data.ISBN}], margin: [0, 0, 0, 11 ]},
            {canvas: [{type: 'line', x1: 0, y1: 0, x2:200, y2: 0, lineWidth: 0.5, lineColor: '#ffffff'}], margin: [0, 0, 0, 15]},
            {text: 'npa.hrc.co.nz', style: ['titleURL']}
          ],
          style:['title','titleReportMeta'],
          width: 200
        },      
        {
          stack: [
            {canvas: [{type: 'line', x1: 0, y1: 0, x2:200, y2: 0, lineWidth: 0.5, lineColor: '#ffffff'}], margin: [0, 0, 0, 15]},
            {image: 'hrc-logo', width: 180, margin: [10, 0, 0, 0]}
          ], 
          width: 200
        }
      ],
      columnGap: 115,      
    }
  ]

}
