// components/PrintTool.js
//React
var React       = require('react')
var Fluxxor         = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

//components
var Icon        = require('components/icon')

//helpers
var _           = require('lodash')
var log         = require('debug')('src:components:printTool')


module.exports = React.createClass({
  propTypes : {
    tool: React.PropTypes.object,
  },
  mixins: [
    Fluxxor.FluxMixin(React),
    StoreWatchMixin('printer')
  ],
  // listen to printer updates
  getStateFromFlux: function () {
    var flux = this.getFlux()
    return {
      status: flux.store('printer').getStatus()            
    }
  },
  
  
  render: function () {
    var tool = this.props.tool
    
    var status = this.state.status
    if (status.code === '' || status.code === 'ready_report') {          
      return (
        <a className='tool-action tool-print-tool' href="#" onClick={this.printPdf(tool.args)} title={tool.label}>
          <Icon name="download"/>
          {tool.label}
        </a>
      )
    } else {
       return (
          <div className='tool-message tool-print-tool'>
            <img alt={'Generating Report'} src="./img/NPA-loading.gif"/>
            {status.message}          
          </div>
      )  
    }
  },
  printPdf: function (args) {
    var flux = this.getFlux()
    return function (e) {
      e.preventDefault()
      flux.actions.export(args.id, args.type)
    }
  }
})
