// components/subjects/recommendation/recommendation.js
var React       = require('react')
var Fluxxor     = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

//components
var Load        = require('components/load')
var EntityTable            = require('components/list/EntityTable')
var CategoryList            = require('components/list/CategoryList')

// bootstrap
var Grid 				= require('react-bootstrap/lib/Grid')
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')

//helpers
var _           = require('lodash')
var marked               = require('utils/parse-markdown')

var log         = require('debug')('src:components:explore:issue')

module.exports = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React),
    StoreWatchMixin('issues', 'actions','recommendations','articles')
  ],

  getStateFromFlux: function () {
    var flux = this.props.flux
    return {
      entitiesLoading: flux.store('recommendations').isLoading(),     
      issuesLoading : flux.store("issues").isLoading(),
      groupsLoading : flux.store("groups").isLoading(),
      treatybodiesLoading : flux.store("treatybodies").isLoading(),
      actionsLoading : flux.store("actions").isLoading()
      
      
    }
  },

   render: function () {
    var flux = this.props.flux
    
    if (this.state.entitiesLoading 
    || this.state.issuesLoading
    || this.state.treatybodiesLoading
    || this.state.groupsLoading 
    || this.state.actionsLoading 
    ) {
      return  (
        <Load />
      )
    } else {
      
       var entity = flux.stores.recommendations.findOne({ id: this.props.params.id })

      return (
      <div className="page-content-issue page-content">
        <Grid className='page-section issue-section-analysis'>
           {this.renderMain(entity)}
        </Grid>
        <Grid className='page-section issue-section-actions'>   
           {this.renderActions(entity)}
        </Grid>        
      </div>  
      )
    }
  },
  renderMain : function (entity){
    var flux = this.props.flux
    return (
      [
        <Row className="header-row">
          <Col md={12}>
            <h2 className="header-type-large">Recommendation</h2>
            <h1 className="header-title">{entity.title}</h1>
          </Col>
        </Row>,
        <Row className="content-main">
          <Col md={8}>
              <div className="recommendation-attributes">
                <table>
                  <tbody>
                    <tr>
                      <td className="label-cell">
                        Recommmendation No
                      </td>
                      <td className="content-cell">
                        {entity.id}
                      </td>
                    </tr>
                    <tr>
                      <td className="label-cell">
                        UPR Session
                      </td>
                      <td className="content-cell">
                        {entity.session}
                      </td>
                    </tr>
                    <tr>
                      <td className="label-cell">
                        Response
                      </td>
                      <td className="content-cell">
                        {entity.response}
                      </td>
                    </tr>
                    <tr>
                      <td className="label-cell">
                        Recommending State
                      </td>
                      <td className="content-cell">
                        {entity.state}
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
          </Col>
          <Col md={4}>  
            <div className='category-list-aside'>
              {this.renderIssues(entity)}
            </div>
            <div className='category-list-aside'>
              {this.renderGroups(entity)}
            </div>
            <div className='category-list-aside'>
              {this.renderTreatybodies(entity)}
            </div>
          </Col>
        </Row>
      ]      
    )
  },
  
  renderActions : function (entity){
    var flux = this.props.flux
    
    var actions =  flux.store('actions').filter(function(item){
      return item.recommendations.split(',').indexOf(entity.id) > -1
    })
    if (actions.lenth === 0){
      return null       
    } else {
      return ([
        <h2>Actions</h2>,
        <EntityTable
          className="action-table"      
          cols = {[
            {type:'no',   title: 'No',      attr:'id'},                        
            {type:'text', title: 'Actions', attr:'title'},
            {type:'links'}
          ]}     
          entities = {actions} />      
      ])
    }
  },
  renderIssues : function (entity){
    var flux = this.props.flux
    
    var items =  flux.store('issues').filter(function(item){
      return entity.issues.split(',').indexOf(item.id) > -1
    })
    return ([
      <h6>{items.length === 1 ? "Issue" : "Issues" }</h6>,
      <CategoryList
        entities = {items} />
    ])
  },
  renderGroups : function (entity){
    var flux = this.props.flux
    
    var items =  flux.store('groups').filter(function(item){
      return entity.groups.split(',').indexOf(item.id) > -1
    })
    return ([
      <h6>{items.length === 1 ? "Population Group" : "Population Groups" }</h6>,
      <CategoryList
        entities = {items} />
    ])
  },
  renderTreatybodies : function (entity){
    var flux = this.props.flux
    
    var items =  flux.store('treatybodies').filter(function(item){
      return entity.treatybodies.split(',').indexOf(item.id) > -1
    })
    return ([
      <h6>{items.length === 1 ? "Treaty Body" : "Treaty Bodies" }</h6>,
      <CategoryList
        entities = {items} />
    ])
  },
 

})
