// components/nav/nav-main.js
//main
var React       = require('react')

// bootstrap
var Navbar          = require('react-bootstrap/lib/Navbar')
var CollapsibleNav  = require('react-bootstrap/lib/CollapsibleNav')
var Nav             = require('react-bootstrap/lib/Nav')
var NavItem         = require('react-bootstrap/lib/NavItem')

//logging
var log       	= require('debug')('components:nav-main')

module.exports = React.createClass({
  initialise : function(){
    // warning: dirty hack to close navbar toggle on route change
    this.route = ''  
    this.param = ''  
    this.expanded = false    
    // end hack
  },
  getInitialState: function() {
    return {
      navExpanded: false
    };
  },            
  render: function () {
    log('props', this.props)
    
    // warning: dirty hack to close navbar toggle on route change  
    //    navExpanded will always override state in NavBar React-Bootstrap component 
   
    var expanded
    if (this.route !== this.props.route 
            || this.param !== this.props.param) {
      expanded = false
      this.route = this.props.route    
      this.param = this.props.param    
      this.expanded = false
    } else {
      expanded = this.state.navExpanded
    }
    //end hack
    
		return (
          <Navbar 
            className='hrc-navbar' 
            brand={this.props.brand}  
            toggleNavKey={0} 
            onToggle={this.handleToggle}
            defaultNavExpanded={false}   
            navExpanded={expanded}
          >
            { this.renderNavbar() }            
          </Navbar>
    )
      
    
  },
  handleToggle : function(){
    log('handleToggle', !this.state.navExpanded)
    // warning: dirty hack to close navbar toggle on route change        
    if (!this.expanded) {
      this.expanded = true
      this.setState({
        navExpanded: true
      });      
    } else {
      this.expanded = false
      this.setState({
        navExpanded: false
      });      
    }
    //end hack
    
        
  },
  renderNavbar : function(){
    return (this.props.isIntro) 
      ? null
      : <CollapsibleNav eventKey={0} expanded={false}>
        {this.renderPagesNavGroup()}
        {this.renderShareNavGroup()}
       </CollapsibleNav>
  },
  renderShareNavGroup : function() {
    var shareClass =  'nav-item-share'
    return (
      <Nav ulClassName={"nav-group nav-group-share"} navbar>
        <NavItem title="Share on Facebook" eventKey={1} href={this.props.facebookShare} className="share-header-social share-header-social-facebook"  target="_blank">
          <div className="icon-social-facebook-sm"/>
        </NavItem>
        <NavItem title="Share on Twitter" eventKey={2} href={this.props.twitterShare} className="share-header-social share-header-social-twitter"  target="_blank">
          <div className="icon-social-twitter-sm"/>
        </NavItem>
        <NavItem title="Open share dialogue" eventKey={3} href='#' className={shareClass} onClick={this.props.toggleShareModal()} >
          Share
        </NavItem>
      </Nav>      
    )    
  },
  renderPagesNavGroup : function() {
    return (
      <Nav ulClassName={"nav-group nav-group-pages"} navbar>
        <NavItem 
            title="Page: About"
            eventKey={1} 
            href='#' 
            active = {this.props.param === 'about'}
            className={'nav-item-about'} 
            onClick={this.props.handleNavClick('about')}>
          About
        </NavItem>
        <NavItem 
            title="Page: Human Rights Articles"
            eventKey={2} 
            href='#' 
            className={'nav-item-articles'} 
            active = {this.props.param === 'articles'}            
            onClick={this.props.handleNavClick('articles')}>
          Human Rights Articles
        </NavItem>
        <NavItem 
            title="Page: Terms"
            eventKey={3} 
            href='#' 
            className={'nav-item-terms'} 
            active = {this.props.param === 'terms'}
            onClick={this.props.handleNavClick('terms')}>
          Terms
        </NavItem>
      </Nav> 
    )    
  },
 
 
})
