// components/main/main.js
//main
var React               = require('react')
var Fluxxor             = require('fluxxor')

var Router        = require('react-router')

var RouteHandler  = Router.RouteHandler

//mixins
var FluxMixin           = Fluxxor.FluxMixin(React)
var StoreWatchMixin     = Fluxxor.StoreWatchMixin

//components
var Header 						= require('../header/')
var Footer            = require('components/footer')
var NavPrimary          = require('components/nav-primary')

// bootstrap
var Grid 				= require('react-bootstrap/lib/Grid')

//helpers
var _                   = require('lodash')


//debugging
var log             = require('debug')('src:components:main:index')

module.exports = React.createClass({

  mixins: [
    FluxMixin,
    StoreWatchMixin(
     
    )
  ],

  propTypes: {
    flux: React.PropTypes.object
  },

  contextTypes: {
    router: React.PropTypes.func
  },

  getStateFromFlux: function () {
    var flux = this.getFlux();
    return {
      
    }
  },

  render: function () {
    var flux = this.getFlux()
    log('rendering main ...')
    
    var navItems = [
      {slug:'issues',route:'issue',title:'Issues'},
      {slug:'groups',route:'group',title:'Population Groups'},
      {slug:'agencies',route:'agency',title:'Government Agencies'},
      {slug:'treatybodies',route:'treatybody',title:'Treaty Bodies'}
    ]
    
    var isIntro = flux.stores.routes.getCurrentRoute() === 'intro'
    
    return (
      <div className='home'>
        <Header
          currentHREF = {window.location.href}
          route = {flux.stores.routes.getCurrentRoute()}          
          title = "New Zealand's National Plan of Action" 
          subtitle = "For the Protection and Promotion of Human Rights"
          flux={flux} />
        <div className="page-wrapper">      
                  
          { (!isIntro)
            ? (
                <NavPrimary
                  flux = {flux}
                  label = { {top:'Explore Actions and Recommendations by',bottom:''} }
                  type = 'exploreby'
                  items = {
                    _.map(navItems,function(item){                
                      var navItem = {}
                      return {
                        slug : item.slug,                  
                        title : item.title,
                        subtitle : '',
                        active : item.slug === flux.stores.routes.getCurrentParams().category 
                          || item.route === flux.stores.routes.getCurrentRoute()
                      }
                    }.bind(this))
                  }
                />
              )
            : ( <div/>)
          }
          <RouteHandler flux={flux}/>
        </div>
        <Footer
          shouldHide = {flux.stores.routes.getCurrentRoute() === 'intro'}
          infoText = "Every care has been taken to ensure the accuracy of this data. Please send any feedback to "
          infoEmail = "infoline@hrc.co.nz"          
          parentSiteText = "Visit our website to learn more about Human Rights in New Zealand"
          parentSiteAnchorText = "www.hrc.co.nz"
          parentSiteURL = "http://hrc.co.nz"
          />
      </div>
    )
  },
})
