// components/main/main.js
//main
var React               = require('react')
var Fluxxor             = require('fluxxor')

var Router        = require('react-router')

var RouteHandler  = Router.RouteHandler

//mixins
var FluxMixin           = Fluxxor.FluxMixin(React)
var StoreWatchMixin     = Fluxxor.StoreWatchMixin

//components
var Header 						= require('../header/')

// bootstrap
var Grid 				= require('react-bootstrap/lib/Grid')

//helpers
var _                   = require('lodash')


//debugging
var log             = require('debug')('src:components:main:index')

module.exports = React.createClass({

  mixins: [
    FluxMixin,
    StoreWatchMixin(
     
    )
  ],

  propTypes: {
    flux: React.PropTypes.object
  },

  contextTypes: {
    router: React.PropTypes.func
  },

  getStateFromFlux: function () {
    var flux = this.getFlux();
    return {
      
    }
  },

  render: function () {
    var flux = this.getFlux()
    log('rendering main ...')

    return (
      <div className='home'>
        <Header
          showTitle = {flux.stores.routes.getCurrentRoute() !== 'intro'}
          title = "New Zealand's National Plan of Action" 
          subtitle = "For the Protection and Promotion oh Human Rights"
          flux={flux} />
        <Grid className='main-content'>
          <RouteHandler flux={flux}/>
        </Grid>
        <footer className='footer container' ></footer>
      </div>
    )
  },
})
