// components/main/main.js
//main
var React               = require('react')
var Fluxxor             = require('fluxxor')

var Router        = require('react-router')

var RouteHandler  = Router.RouteHandler

//mixins
var FluxMixin           = Fluxxor.FluxMixin(React)
var StoreWatchMixin     = Fluxxor.StoreWatchMixin

//components
var Header 						= require('../header/')
var NavPrimary          = require('components/nav-primary')

// bootstrap
var Grid 				= require('react-bootstrap/lib/Grid')

//helpers
var _                   = require('lodash')


//debugging
var log             = require('debug')('src:components:main:index')

module.exports = React.createClass({

  mixins: [
    FluxMixin,
    StoreWatchMixin(
     
    )
  ],

  propTypes: {
    flux: React.PropTypes.object
  },

  contextTypes: {
    router: React.PropTypes.func
  },

  getStateFromFlux: function () {
    var flux = this.getFlux();
    return {
      
    }
  },

  render: function () {
    var flux = this.getFlux()
    log('rendering main ...')
    
    var navItems = [
      {slug:'issues',title:'Issues'},
      {slug:'groups',title:'Population Groups'},
      {slug:'agencies',title:'Government Agencies'},
      {slug:'treatybodies',title:'Treaty Bodies'}
    ]
    
    var isIntro = flux.stores.routes.getCurrentRoute() !== 'intro'
    
    return (
      <div className='home'>
        <Header
          showTitle = {isIntro}
          title = "New Zealand's National Plan of Action" 
          subtitle = "For the Protection and Promotion oh Human Rights"
          flux={flux} />
          { (isIntro)
            ? (
                <NavPrimary
                  flux = {flux}
                  label = { {top:'Explore Actions and Recommendations by',bottom:''} }
                  type = 'exploreby'
                  items = {
                    _.map(navItems,function(item){                
                      var navItem = {}
                      return {
                        slug : item.slug,                  
                        title : item.title,
                        subtitle : '',
                        active : item.slug === flux.stores.routes.getCurrentParams().category
                      }
                    }.bind(this))
                  }
                />
              )
            : ( <div/>)
          }
        <Grid className='main-content'>
          <RouteHandler flux={flux}/>
        </Grid>
        <footer className='footer container' ></footer>
      </div>
    )
  },
})
