// printPdf/printPdf.js
// helpers
var _ = require('lodash')
var $script  = require('scriptjs')
var log     = require('debug')('src:node_modules:printPdf')

// util
var parseHTML = require('./parse-html')
var getImageDataUrl = require('./getImageDataUrl')

// layouts
var titleLayout = require('./layouts/title-layout')
var actionLayout = require('./layouts/action-layout')
var categoryLayout = require('./layouts/category-layout')
var actionsTableLayout = require('./layouts/actions-table-layout')
var recommendationsTableLayout = require('./layouts/recommendations-table-layout')
//var allLayout = require('utils/print/all-layout')

module.exports = function (report) {  
  // define styles
  var defaultStyles = {
    color: '#0A262D',
    columnGap: 20,
    fontSize: 8.5,
    lineHeight: 1.2,
    margin:[0,0,0,5]
  }
  var styles = { 
    title :{
      color:'#ffffff'
    },
    titleHeaderTitle : {
      bold:true,
      fontSize:12
    },
    titleHeaderSubtitle : {
      fontSize:10
    },
    titleReportType : {
      bold:true,
      fontSize:12,
      color:'#029BE5'
    },
    titleReportSupTitle : {
      fontSize:10,
      bold:true
    },
    titleReportTitle : {
      fontSize:24,
      lineHeight:1.1
    },
    titleInReport : {
      fontSize:10,
      bold:true
    },
    titleReportMeta :{
      color:'#ffffff',
      fontSize:8
    },
    titleURL :{
      fontSize:20,
      bold:true
    },
    pageIdTable : {
      color:'#ffffff',
      alignment:'center'
    },
    pageIdTableCellType : {
      fontSize:11,
      bold : true,
      lineHeight: 0.7,      
      fillColor: '#69A41E',
      margin : [0,11,0,0]              
    },
    pageIdTableCellId : {
      fontSize:21,
      bold : true,
      lineHeight: 1, 
      fillColor: '#69A41E',      
      margin : [0,0,0,9]
    },
    pageIdTableCellTypeRec : {    
      fillColor: '#029BE5',
      margin : [0,22,0,22] 
    },
    itemSupTitle : {
      fontSize: 10,
      color : '#869497',
      bold : true,
      margin: [0, 53, 0, 0]
    },
    itemTitleAction : {
      fontSize: 16,
      bold : false,
      margin: [0, 10, 0, 22]   
    },
    itemTitleRecTable : {
      fontSize: 14,
      bold : true,
      margin: [0, 10, 0, 22]   
    },
    itemContentSectionTitle : {
      bold: true,
      margin: [0, 10, 0, 5]
    },
    sideTableCell : {      
      fillColor:'#F5F7F0',
      margin : [10,0]
    },
    sideTableCellTitle : {
      bold: true,
      margin : [10,10]
    },
    sideTable : {
      margin:[0,0,0,10]
    },
    smartIcon : {
      width: 25
    },
    footerTitle:{
      color: '#869497',
      fontSize: 7
    },
    footerPageNo:{
      color: '#869497',
      fontSize: 16
    },
    tableHeader :{
      color: '#869497',
      fontSize: 7
    },
    tableKey :{      
      fontSize: 6
    },
    cellId:{
      margin:[0,2,0,0],
      fontSize:12
    },
    cellTitle:{      
      margin:[0,2],
      fontSize:8
    },
    cellIcon:{
      margin:[5,2],
    },
    paragraph:{margin:[0,0,0,10]},
    grey: {color: '#869497'},    
    blue: {color: '#029BE5'},
    green: {color: '#69A41E'},    
    white: {color: '#ffffff'},    
    bold: {bold: true}    
  }
  
  //define global design elements
  var elements = {
    lines : {
      contentBorderTop : {
        canvas: [{ type: 'line', x1: 0, y1:0, x2: 320, y2: 0, lineWidth: 0.5, lineColor :'#B6BDBF' }],
        margin: [0, 0, 0, 10]
      },
      aside : {
        canvas: [{ type: 'line', x1: 0, y1: 0, x2: 123, y2: 0, lineWidth: 0.5, lineColor :'#fff' }]
      }
    }
  }
  //define images
  var images = {
     defaultImages : [
       { name : 'hrc-logo' },
       { name : 'hrc-logo-icon' },
     ],
     action : [
       { name : 'smart-s-pass-active-bg' },
       { name : 'smart-s-fail-active-bg' },
       { name : 'smart-m-pass-active-bg' },
       { name : 'smart-m-fail-active-bg' },
       { name : 'smart-a-pass-active-bg' },
       { name : 'smart-a-fail-active-bg' },
       { name : 'smart-r-pass-active-bg' },
       { name : 'smart-r-fail-active-bg' },
       { name : 'smart-t-pass-active-bg' },
       { name : 'smart-t-fail-active-bg' }
     ],
     recommendationTable: [
       { name : 'pass-active' },
       { name : 'fail-active' },      
     ],
     agency: [
       { name : 'cat-agencies-white' }
     ],
     group : [
       { name : 'cat-groups-white' }
     ],
     issue : [
       { name : 'cat-issues-white' }
     ],
     actionTable : [
       { name : 'smart-s-pass-active' },
       { name : 'smart-s-fail-active' },
       { name : 'smart-m-pass-active' },
       { name : 'smart-m-fail-active' },
       { name : 'smart-a-pass-active' },
       { name : 'smart-a-fail-active' },
       { name : 'smart-r-pass-active' },
       { name : 'smart-r-fail-active' },
       { name : 'smart-t-pass-active' },
       { name : 'smart-t-fail-active' }       
     ]
   }
  
  var content = [],   
    imagesToLoad,
    sectionStartPage = 1,    
    currentSection = 'X'

  // prepare report data based on report type
  switch (report.type) {
    // single action report
    case 'action' :  
      // action layout
      // action-layout
      // - title-layout
      // - A
      //   - action-content-layout
      // - B
      //   - recommendations-table-layout
      
      content.push({
        stack: titleLayout(report.titleData,report.sections,elements),
        pageBreak:'after',
        footerMode: '0' 
      })
      content.push({ 
        stack: actionLayout(report.data.actionData, elements),
        pageBreak:'after',
        footerMode: 'A'    
      })
      content.push({ 
        stack: recommendationsTableLayout(report.data.recommendationsTable,report.type,elements),
        footerMode: 'B'
      })
      imagesToLoad = images.defaultImages.concat(images.action).concat(images.recommendationTable)
      break

    case 'issue' :
    case 'group' :
    case 'agency' :            
      
      // category layout
      // category-layout
      // - title-layout
      // - A
      //   - category-content-layout
      //   - action-table-layout
      //   - recommendations-table-layout
      // - B
      //   - each action
      //     - action-content-layout      
      content.push({
        stack: titleLayout(report.titleData,report.sections,elements),
        pageBreak:'after',
        footerMode: '0' 
      })
      content.push({ 
        stack: categoryLayout(report.data.categoryData, report.type, false, elements),        
        pageBreak:'after',
        footerMode: 'A',         
      })     
      content.push({ 
        stack: actionsTableLayout(
                report.data.categoryData.actionsTable,
                report.type,
                elements),
        pageBreak:'after',
        footerMode: 'A'
      }) 
      if (report.type !== 'agency') {
        content.push({ 
          stack: recommendationsTableLayout(
                  report.data.categoryData.recommendationsTable,
                  report.type,
                  elements),
          pageBreak:'after',
          footerMode: 'A'
        })
      }
      _.each(report.data.actionsDataList,function(actionData,i){
        content.push({ 
          stack: actionLayout(actionData, elements),
          pageBreak: i < report.data.actionsDataList.length-1 ? 'after' : 'none',
          footerMode: 'B'        
        })        
      })
      
      imagesToLoad = images.defaultImages
              .concat(images[report.type])
              .concat(images.actionTable)
              .concat(images.recommendationTable)
              .concat(images.action)
      break
    case 'all' :     
      // All layout
      // all-layout
      // - title-layout
      // - A
      //   - each issue: 
      //      - category-content-layout
      //      - action-table-layout
      //   - each group: 
      //      - category-content-layout
      //      - action-table-layout
      //   - each agency:
      //      - category-content-layout
      //      - action-table-layout   
      // - B
      //   - each action: 
      //     - action-content-layout        
      // - C
      //     - recommendations-table-layout
      
        
      imagesToLoad = _.flatten(_.values(images))
      break
  }

    
  // the document definition object
  var docDefinition = {
    images : {},
    content: content,
    styles: styles,
    defaultStyle: defaultStyles,
    pageSize: 'A4', // A4: [595.28, 841.89]
    pageMargins: [ 40.14, 40, 40.14, 80 ],
    background: function (currentPage) {
      // title page 
      if (currentPage === 1){ // turn off temporarily
        return {canvas: [
          {
            type: 'polyline',
            lineWidth: 0,
            color: '#029BE5',
            lineColor: '',
            points: [{x: 0, y: 0}, {x: 0, y: 841.89}, {x: 595.28, y: 841.89}, {x: 595.28, y: 170}, {x: 425.28, y: 0}, {x: 0, y: 0}]
          }
        ]}
      }
    },
    // footerMode : section
    footer: function (currentPage, pageCount, footerMode) {
      if (footerMode !== '0') {
        var pageNo
         
        // determine section change and remember section start page
        if (typeof footerMode !== 'undefined' && footerMode !== currentSection) {
          sectionStartPage = currentPage
          currentSection = footerMode
        }            
        pageNo = currentPage - sectionStartPage + 1          

        return {
          stack : [
            { 
              canvas: [{ type: 'line', x1: 0, y1: 0, x2: 515, y2: 0, lineColor: '#B6BDBF', lineWidth: 0.5 }],
              margin:[0,10]
            },
            {
              columns : [
                {                  
                  image: 'hrc-logo-icon',
                  width:40,
                  margin:[0,0,10,0]                  
                },
                {
                  stack : [
                    {text : 'New Zealand National Plan of Action'},
                    {text : report.footer.title}
                  ],
                  style:'footerTitle',
                  margin:[0,0,10,0]                  
                },
                {
                  text: currentSection + ' | ' + pageNo, 
                  alignment: 'right',
                  width:50,
                  style:'footerPageNo'
                }
              ]
            }
          ],
          margin:[40,20]
        }
        
      }
    }
  }         
        
        
        
  // wait for images to be loaded before printing pdf
  
  
  
  loadImages (imagesToLoad)
  
  waitForImages(
    imagesToLoad,
    function(){      
      log('finished loading pdf icons')
      $script('js/pdfmake.min.js',function(){
        $script('js/vfs_fonts.js', function(){              
          log('finished loading pdf scripts and fonts')
          pdfMake.createPdf(docDefinition).download(report.filename)          
        })
      })
    },
    function (e) {
      log(e.message)
    }
  )     
  
  function loadImages(images){
    var imgPath = 'img/pdf-icons/hrc-icons_'
    _.each(images,function(img){
      getImageDataUrl(imgPath + img.name + '.png',
      function(datauri){
        img.datauri = datauri
        docDefinition.images[img.name] = datauri
      },
      function (e) {
        img.e = e
      })
    })
  }
  
  function waitForImages(images,onSuccess, onError){
    
    var success = true
    var error = false
    
    _.each(images,function(img){
      success = success && typeof img.datauri !== 'undefined'
      error = error || typeof img.e !== 'undefined'
    })
    
    if (success){
      onSuccess();
    } else if (error){
      onError(img.e);
    } else {
      setTimeout(function(){
        waitForImages(images,onSuccess, onError);
      },250);
    }
  }    
}
