// printPdf/printPdf.js
// helpers
var _ = require('lodash')
var $script  = require('scriptjs')

// util
var parseHTML = require('./parse-html')
var getImageDataUrl = require('./getImageDataUrl')

// layouts
var titleLayout = require('./layouts/title-layout')
var actionLayout = require('./layouts/action-layout')
var recommendationsTableLayout = require('./layouts/recommendations-table-layout')
//var categoryLayout = require('utils/print/category-layout')
//var categoryContentLayout = require('utils/print/categorycontent-layout')
//var allLayout = require('utils/print/all-layout')

module.exports = function (report) {  
  // define styles
  var defaultStyles = {
    color: '#0A262D',
    columnGap: 20,
    fontSize: 8.5,
    lineHeight: 1.2,
    margin:[0,0,0,5]
  }
  var styles = {
    pageIdTable : {
      color:'#ffffff',
      alignment:'center'
    },
    pageIdTableCellType : {
      fontSize:11,
      bold : true,
      lineHeight: 0.7,      
      fillColor: '#69A41E',
      margin : [0,11,0,0]              
    },
    pageIdTableCellId : {
      fontSize:21,
      bold : true,
      lineHeight: 1, 
      fillColor: '#69A41E',      
      margin : [0,0,0,9]
    },
    pageIdTableCellTypeRec : {    
      fillColor: '#029BE5',
      margin : [0,22,0,22] 
    },
    itemSupTitle : {
      fontSize: 10,
      color : '#869497',
      bold : true,
      margin: [0, 53, 0, 0]
    },
    itemTitleAction : {
      fontSize: 16,
      bold : false,
      margin: [0, 10, 0, 22]   
    },
    itemTitleRecTable : {
      fontSize: 14,
      bold : true,
      margin: [0, 10, 0, 22]   
    },
    itemContentSectionTitle : {
      bold: true,
      margin: [0, 10, 0, 5]
    },
    sideTableCell : {      
      fillColor:'#F5F7F0',
      margin : [10,0]
    },
    sideTableCellTitle : {
      bold: true,
      margin : [10,10]
    },
    sideTable : {
      margin:[0,10]
    },
    smartIcon : {
      width: 25
    },
    footerTitle:{
      color: '#869497',
      fontSize: 7
    },
    footerPageNo:{
      color: '#869497',
      fontSize: 16
    },
    tableHeader :{
      color: '#869497',
      fontSize: 7
    },
    tableKey :{      
      fontSize: 6
    },
    cellId:{
      margin:[0,2,0,0],
      fontSize:12
    },
    cellTitle:{      
      margin:[0,2],
      fontSize:8
    },
    cellIcon:{
      margin:[5,2],
    },
    paragraph:{margin:[0,0,0,10]},
    grey: {color: '#869497'},    
    blue: {color: '#029BE5'},
    green: {color: '#69A41E'},    
    white: {color: '#ffffff'},    
    bold: {bold: true}    
  }
  
  //define global design elements
  var elements = {
    lines : {
      contentBorderTop : {
        canvas: [{ type: 'line', x1: 0, y1:0, x2: 320, y2: 0, lineWidth: 0.5, lineColor :'#B6BDBF' }],
        margin: [0, 0, 0, 10]
      },
      aside : {
        canvas: [{ type: 'line', x1: 0, y1: 0, x2: 123, y2: 0, lineWidth: 0.5, lineColor :'#fff' }]
      }
    }
  }
  //define images
  var images = {
     defaultImages : [
       { name : 'hrc-logo' },
       { name : 'hrc-logo-icon' },
     ],
     action : [
       { name : 'pass-active' },
       { name : 'fail-active' },
       { name : 'smart-s-pass-active-bg' },
       { name : 'smart-s-fail-active-bg' },
       { name : 'smart-m-pass-active-bg' },
       { name : 'smart-m-fail-active-bg' },
       { name : 'smart-a-pass-active-bg' },
       { name : 'smart-a-fail-active-bg' },
       { name : 'smart-r-pass-active-bg' },
       { name : 'smart-r-fail-active-bg' },
       { name : 'smart-t-pass-active-bg' },
       { name : 'smart-t-fail-active-bg' }
     ]
   }
  
  var content = [],   
    sectionStartPage = 1,    
    currentSection = 'X'

  // prepare report data based on report type
  switch (report.type) {
    // single action report
    case 'action' :  
      // action layout
      // action-layout
      // - title-layout
      // - A
      //   - action-content-layout
      // - B
      //   - recommendations-table-layout
      
      content.push({
        stack: titleLayout(report.titleData,report.sections,elements),
        pageBreak:'after',
        footerMode: '0' 
      })
      content.push({ 
        stack: actionLayout(report.data.actionData, elements),
        footerMode: 'A',
        pageBreak:'after'         
      })
      content.push({ 
        stack: recommendationsTableLayout(report.data.recommendationsTable,report.type,elements),
        footerMode: 'B'
      })

      break
      
      
      // category layout
      // category-layout
      // - title-layout
      // - A
      //   - category-content-layout
      //   - action-table-layout
      //   - recommendations-table-layout
      // - B
      //   - each action
      //     - action-content-layout
      // 
      // 
      // 
        // issue category report
//        case 'issue' :
//          content = this.getCategory(payload, category, 'issues')
//          reportTitle = 'New Zealand’s National Plan of Action | Category report | Issue: ' + category.title
//          break
//        // group category report
//        case 'group' :
//          content = this.getCategory(payload, category, 'groups')
//          reportTitle = 'New Zealand’s National Plan of Action | Category report | Group: ' + category.title
//          break
//        // agency category report
//        case 'agency' :
//          content = this.getCategory(payload, category, 'agencies')
//          reportTitle = 'New Zealand’s National Plan of Action | Category report | Agency: ' + category.title
//          break
//        // full report
//        
//        
      // All layout
      // all-layout
      // - title-layout
      // - A
      //   - each issue: 
      //      - category-content-layout
      //      - action-table-layout
      //   - each group: 
      //      - category-content-layout
      //      - action-table-layout
      //   - each agency:
      //      - category-content-layout
      //      - action-table-layout   
      // - B
      //   - each action: 
      //     - action-content-layout        
      // - C
      //     - recommendations-table-layout
      
//        
//        case 'all' :
//
//          content = this.getAll(payload, relatedEntities, allAgencies, allGroups, allIssues)
//          reportTitle = 'New Zealand’s National Plan of Action | Full report'
//          break
//        // custom filtered report
//        case 'filtered' :
//          break
  }

    
  // the document definition object
  var docDefinition = {
    images : {},
    content: content,
    styles: styles,
    defaultStyle: defaultStyles,
    pageSize: 'A4', // A4: [595.28, 841.89]
    pageMargins: [ 40.14, 40, 40.14, 80 ],
    background: function (currentPage) {
      // title page if not single action report
      if (currentPage === 1){
        return {canvas: [
          {
            type: 'polyline',
            lineWidth: 2,
            color: '#029BE5',
            lineColor: '',
            points: [{x: 40, y: 40}, {x: 40, y: 800}, {x: 560, y: 800}, {x: 560, y: 90}, {x: 510, y: 40}, {x: 40, y: 40}]
          }
        ]}
      }
    },
    footer: function (currentPage, pageCount, footerMode) {
      if (footerMode !== '0') {
        var pageNo
         
        // determine section change and remember section start page
        if (footerMode !== currentSection) {
          sectionStartPage = currentPage
          currentSection = footerMode
        }            
        pageNo = currentPage - sectionStartPage + 1          

        return {
          stack : [
            { 
              canvas: [{ type: 'line', x1: 0, y1: 0, x2: 515, y2: 0, lineColor: '#B6BDBF', lineWidth: 0.5 }],
              margin:[0,10]
            },
            {
              columns : [
                {                  
                  image: 'hrc-logo-icon',
                  width:40,
                  margin:[0,0,10,0]                  
                },
                {
                  stack : [
                    {text : 'New Zealand National Plan of Action'},
                    {text : report.footer.title}
                  ],
                  style:'footerTitle',
                  margin:[0,5,10,0]                  
                },
                {
                  text: footerMode + ' | ' + pageNo, 
                  alignment: 'right',
                  width:50,
                  style:'footerPageNo'
                }
              ]
            }
          ],
          margin:[40,20]
        }
        
      }
    }
  }         
        
        
        
  // wait for images to be loaded before printing pdf
  
  var imagesToLoad = images.defaultImages.concat(images[report.type] )
  
  loadImages (imagesToLoad)
  
  waitForImages(
    imagesToLoad,
    function(){      
      $script('js/pdfmake.min.js',function(){
        $script('js/vfs_fonts.js', function(){              
          pdfMake.createPdf(docDefinition).download(report.filename)
        })
      })
    },
    function (e) {
      console.log(e.message)
    }
  )     
  
  function loadImages(images){
    var imgPath = 'img/pdf-icons/hrc-icons_'
    _.each(images,function(img){
      getImageDataUrl(imgPath + img.name + '.png',
      function(datauri){
        img.datauri = datauri
        docDefinition.images[img.name] = datauri
      },
      function (e) {
        img.e = e
      })
    })
  }
  
  function waitForImages(images,onSuccess, onError){
    
    var success = true
    var error = false
    
    _.each(images,function(img){
      success = success && typeof img.datauri !== 'undefined'
      error = error || typeof img.e !== 'undefined'
    })
    
    if (success){
      onSuccess();
    } else if (error){
      onError(img.e);
    } else {
      setTimeout(function(){
        waitForImages(images,onSuccess, onError);
      },250);
    }
  }    
}
