// actions/actions.js
// helpers
var log       = require('debug')('actions:index')


// define routes
var c = {
  ROUTE: {
    TRANSITION: "ROUTE:TRANSITION",
    GOBACK: "ROUTE:GOBACK",
  },
  PRINT: "PRINT"
};

var methods = {
  // select single entity
  entity: {
    select : function (type,id) {
      this.dispatch(c.ROUTE.TRANSITION, { route: type, params: { id: id } });
    }
  },
  // entity overview by category
  exploreby:  function (category) {
      this.dispatch(c.ROUTE.TRANSITION, { route: 'explore-by', params: { id: category } });

  },
  // navigate to page
  navPage: function (page){
    log('navPage', page)
    if (page === 'home') {
      this.dispatch(c.ROUTE.TRANSITION, { route: '/' })
    } else {
      this.dispatch(c.ROUTE.TRANSITION, { route: '/page/' + page })
    }

  },
  // navigate back
  navigateBack : function () {
    this.dispatch(c.ROUTE.GOBACK,{})
  },

  export: function (id, type) {
    this.dispatch(c.PRINT, {id: id, type: type})
  }
};

module.exports = {
  methods: methods,
  constants: c
};
