// actions/actions.js
// helpers
var log       = require('debug')('actions:index')
var createPdf = require('pdfmake-browserified')

// define routes
var c = {
  ROUTE: {
    TRANSITION: "ROUTE:TRANSITION",
    GOBACK: "ROUTE:GOBACK",
  }
};

var methods = {
  // select single entity
  entity: {
    select : function (type,id) {
      this.dispatch(c.ROUTE.TRANSITION, { route: type, params: { id: id } });
    }
  },
  // entity overview by category
  exploreby:  function (category) {
      this.dispatch(c.ROUTE.TRANSITION, { route: 'explore-by', params: { id: category } });

  },
  // navigate to page
  navPage: function (page){
    log('navPage', page)
    if (page === 'home') {
      this.dispatch(c.ROUTE.TRANSITION, { route: '/' })
    } else {
      this.dispatch(c.ROUTE.TRANSITION, { route: '/page/' + page })
    }

  },
  // navigate back
  navigateBack : function () {
    this.dispatch(c.ROUTE.GOBACK,{})
  },

  exportAction: function (entity) {
    var dd = {
      content: [
        { text: entity.title, style: 'header1', margin: [0, 20] },
        { text: 'DESCRIPTION', style: 'header2', margin: [0, 10] },
        {
          columns: [
            {
              stack: [
                { text: entity.description },
                { text: 'DESIRED OUTCOME', style: 'header2', margin: [0, 10] },
                { text: entity.desiredoutcome },
                { text: 'MEASURES', style: 'header2', margin: [0, 10] },
                { text: entity.measures },
                { text: 'TARGET DATE', style: 'header2', margin: [0, 10] },
                { text: entity.targetdate }
              ],
              width: '80%'
            },
            [
              { text: 'Issue', style: 'header3', margin: [0, 10] },
              { text: entity.issues, style: 'blue' },
              { text: 'Population Groups', style: 'header3', margin: [0, 10] },
              { text: entity.groups, style: 'blue' },
              { text: 'Government Agency', style: 'header3', margin: [0, 10] },
              { text: entity.agencies, style: 'blue' }
            ]
          ]
        }
      ],

      styles: {
        header1: {
          fontSize: 22,
          bold: true
        },
        header2: {
          fontSize: 15,
          bold: true,
          color: '#7696A5'
        },
        header3: {
          fontSize: 12,
          bold: true
        },
        blue: {
          color: '#008399'
        }
      },
      defaultStyle: {
        columnGap: 20
      }
    }
    createPdf(dd).open()
  }
};

module.exports = {
  methods: methods,
  constants: c
};
