// components/explore/treatybody/treatybody.js
var React           = require('react')
var Fluxxor         = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

//components

// bootstrap
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')

//helpers

var _               = require('lodash')

var log       = require('debug')('src:components:explore:treatybody')

module.exports = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React),
    StoreWatchMixin('treatybodies', 'recommendations')
  ],

  getStateFromFlux: function () {
    var flux = this.props.flux
    return {
      entitiesLoading: flux.store('treatybodies').isLoading(),
      recommendationsLoading : flux.store("recommendations").isLoading()      
    }
  },


  render: function () {
    var flux = this.props.flux

    return (
      <Row>
        { this.renderMain() }
      </Row>
    )
  },
  renderMain : function (){
    var flux = this.props.flux
    var entity = flux.stores.treatybodies.findOne({ id: this.props.params.id })
log('treatybody',entity)
    return (
      <Col md={12}>
        <p>Treatybody </p>
        <p>
          { (this.state.entitiesLoading) ? 'loading' : entity.title }
        </p>
      </Col>
    )
  },


})
