// components/explore/explore.js
var React           = require('react')
var Fluxxor         = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

//components

// bootstrap
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')

//helpers

var _               = require('lodash')

var log       = require('debug')('src:components:issue')

module.exports = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React),
    StoreWatchMixin('issues', 'groups', 'agencies','treatybodies','recommendations','actions')
  ],

  getStateFromFlux: function () {
    var flux = this.props.flux
    return {
      issuesLoading: flux.store('issues').isLoading(),
      groupsLoading: flux.store('groups').isLoading(),
      agenciesLoading: flux.store('agencies').isLoading(),
      treatybodiesLoading: flux.store('treatybodies').isLoading(),
      actionsLoading : flux.store("actions").isLoading(),
      recsLoading : flux.store("recommendations").isLoading()
      
      
    }
  },

  render: function () {
    var flux = this.props.flux

      
    return (
      <Row>
        { this.renderMain() }
      </Row>
   
    )
  },
  renderMain : function (){
    var flux = this.props.flux

    return (
      <Col md={12}>
        <p>Explore by</p>
        <p>
        {this.props.params.category}
        </p>
      </Col>
    )
  },
 

})
