// components/pages/intro/intro.js
var React         = require('react')
var Fluxxor       = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

var Load        = require('components/load')


//helpers
var _         = require('lodash')

var Grid 				= require('react-bootstrap/lib/Grid')
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')

var log       = require('debug')('src:components:intro')


module.exports = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React),
    StoreWatchMixin('actions','recommendations')    
  ],
  getStateFromFlux: function () {
    var flux = this.getFlux()    
    return {
      recsLoading : flux.store("recommendations").isLoading(),
      actionsLoading : flux.store("actions").isLoading(),
    }
  },
  render: function () {

    log('props, state', this.props, this.state)
    var flux = this.getFlux()    
    var SelectView
    var navItems = [
      {slug:'issues',title:'Issue'},
      {slug:'groups',title:'Population Group'},
      {slug:'agencies',title:'Government Agency'},
      {slug:'treatybodies',title:'UN Treaty Body'}
    ]

      

      SelectView = ( [
      <Grid key={0} className="intro-explore-hint">                      
        <h4 className="serif intro-explore">Explore actions and recommendations by</h4>
      </Grid>,
      <div key={1} className="intro-issue-select nav-primary">  
        <Grid>              
          <ul className="list-inline nav-primary-list">
            {
              _.map(navItems,function (item, i, items) {
                var width = 1/items.length * 100 + '%'
                var className = 'nav-primary-li intro-li'
                return (
                  <li key={i} className={className}  style={{width:width}}>
                    <a
                      className='btn'
                      href='#'
                      onClick={this.handleCategoryClick(item.slug)}>
                      <span className={'icon-'+ item.slug}></span>
                      <div className='title'>{item.title}</div>
                    </a>
                  </li>
                )
              },this)
            }
          </ul>
        </Grid>
      </div>,
      <Grid key={2} >
        <div className="intro-read-more">
          <p>or learn more about the UPR process <a href="#" onClick={this.handleNavClick('about')}>here</a></p>
        </div>
      </Grid>]  
      )
    
    return (
      <div className='intro text-center'>
        <Grid>
          <Row className="intro-content">
            <Col sm={10} smPush={1}>
              <h1 className='serif'>
                New Zealand Human Rights 
                <br/> 
                National Plan of Action 2015 &ndash; 2019 
              </h1>
              <p>
                This National Plan sets out the actions the Government is taking 
                as a result of the commitments it made through its second Universal Periodic Review (UPR)
                before the United Nations Human Rights Council in 2014
              </p>
            </Col>
          </Row>
        </Grid>
        {this.renderStats()}

       {SelectView}
      </div>
    )
  },
  renderStats : function (){
    if ( this.state.recsLoading 
      || this.state.actionsLoading 
    ) {
      return  (
        <Load />
      )
    } else {    
      var flux = this.getFlux()    
      var no_submissions = 54
      var no_recs = flux.store('recommendations').entities().length
      var no_actions = flux.store('actions').entities().length
      
      return(
            <Grid>
              <Row className="intro-stats hidden-xs">
                <Col sm={10} smPush={1}>
                  <ul className="list-inline">
                    <li className="submission-stat intro-stat-item">
                      <div className="stat-large">{no_submissions}</div>
                      <div className="stat-title">Submissions to the UPR</div>
                      <div className="stat-content">by 250+ civil society individuals and groups</div>
                    </li>
                    <li className="intro-stat-divider">
                      <span className="icon-arrow-tall"/>
                    </li>
                    <li className="recs-stat intro-stat-item">
                      <div className="stat-large">{no_recs}</div>
                      <div className="stat-title">UPR Recommendations</div>
                      <div className="stat-content">of which 121 were accepted by the Government</div>
                    </li>
                    <li className="intro-stat-divider">
                      <span className="icon-arrow-tall"/>
                    </li>            
                    <li className="action-stat intro-stat-item">
                      <div className="stat-large">{no_actions}</div>
                      <div className="stat-title">Government Actions</div>
                      <div className="stat-content">address the UPR recommendations</div>
                    </li>
                  </ul>
                </Col>
              </Row>
            </Grid>
      )
    }
  },
  handleCategoryClick: function (slug) {   
    return _.bind(function (e) {
      e.preventDefault()
      this.getFlux().actions.exploreby(slug)
    },this)
  },
  handleNavClick: function (page) {
    var flux = this.props.flux

    return function (e) {
      e.preventDefault()
      flux.actions.navPage(page)
    }
  },          
})
