// components/views/all/all.js
var React       = require('react')
var Fluxxor     = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

//components
var Load        = require('components/load')
var PageNotFound= require('components/page-not-found')

//var Filters = require('components/filters/filters')

// TODO TEMP
var NavTop        = require('components/nav/nav-categories')
var EntityTableWrap            = require('components/table/EntityTableWrap')

// bootstrap
var Grid 				= require('react-bootstrap/lib/Grid')
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')

//util
var referencing = require('utils/referencing')
var getColumns = require('utils/getColumns')
var checkSMART = require('utils/checkSMART')

//helpers
var _           = require('lodash')
var moment      = require('moment')
var log         = require('debug')('src:components:filter')

module.exports = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React),
    StoreWatchMixin('actions','recommendations','issues','groups','agencies','treatybodies')
  ],
  propTypes : {
    params: React.PropTypes.object,
    query: React.PropTypes.object
  },
  getInitialState : function(){
    return {
      filtergroups : {
        cat:      {title:'Categories',type:'multiselect'},
        rec:      {title:'Recommendations',type:'multiselect'},
        action:   {title:'Actions',type:'multiselect'},        
        date:     {title:'Target Date',type:'dateselect'},
        smart:    {title:'SMART criteria',type:'icon',states:['Any','Satisfied','Not satisfied']},        
        response: {title:'Government Response',type:'icon',states:['Any','Accepted','Not Accepted']}
      },
      types : {
        actions : [
          {by:'issue',type:'entities',attr:'issues',relation:'to',filtergroup:'cat',filterlabel:'Issues'},
          {by:'group',type:'entities',attr:'groups',relation:'to',filtergroup:'cat',filterlabel:'Groups'},
          {by:'agency',type:'entities',attr:'agencies',relation:'to',filtergroup:'cat',filterlabel:'Agencies'},
          {by:'recommendation',type:'entities',attr:'recommendations',relation:'to',filtergroup:'rec',filterlabel:'Recommendations'},
          {by:'specific',type:'smart',attr:'isspecific',filtergroup:'smart',filterlabel:'Specific'},
          {by:'measurable',type:'smart',attr:'ismeasurable',filtergroup:'smart',filterlabel:'Measurable'},
          {by:'assignable',type:'smart',attr:'isassignable',filtergroup:'smart',filterlabel:'Assignable'},
          {by:'resultoriented',type:'smart',attr:'isresultoriented',filtergroup:'smart',filterlabel:'Result-oriented'},
          {by:'timebound',type:'smart',attr:'istimebound',filtergroup:'smart',filterlabel:'Time-bound'},
          {by:'target-date-after',type:'date',attr:'targetdate', relation:'after',filtergroup:'date',filterlabel:'After'},
          {by:'target-date-before',type:'date',attr:'targetdate', relation:'before',filtergroup:'date',filterlabel:'Before'}
        ],
        recommendations : [
          {by:'issue',type:'entities',attr:'issues',relation:'to',filtergroup:'cat',filterlabel:'Issues'},
          {by:'group',type:'entities',attr:'groups',relation:'to',filtergroup:'cat',filterlabel:'Groups'},
          {by:'treatybody',type:'entities',attr:'treatybodies',relation:'to',filtergroup:'cat',filterlabel:'Treaty Bodies'},
          {by:'action',type:'entities',attr:'recommendations',relation:'from',from:'actions',filtergroup:'action',filterlabel:'Actions'},
          {by:'accepted',type:'response',attr:'response',filtergroup:'response',filterlabel:'Accepted'},
          {by:'rejected',type:'response',attr:'response',filtergroup:'response',filterlabel:'Not Accepted'}
        ]          
      }
    }
  },
  getStateFromFlux: function () {
    var flux = this.getFlux()
    return {
      loading: flux.store('actions').isLoading() 
           ||  flux.store("recommendations").isLoading()
    }
  },

  render: function () {
    log('props',this.props)
     
    var flux = this.getFlux()
    
    var type = this.props.params.id
    var filters = _.clone(this.props.query)
    
    // check if type is allowed
    if (Object.keys(this.state.types).indexOf(type) === -1)  {
              
      return ( 
        <div>
          <NavTop />
          <PageNotFound />
        </div>
      )
    } else if (this.state.loading) {
      return (
        <div>
          <NavTop 
            active = 'actions'
          />  
          <Load/>
        </div>
      )
    } else {
      
            
      
      typeFilters = this.state.types[type]
      
      // collect conditions
      
      var conditions = []
      // for each filter from filters
      _.each(filters,function(filter,key){
        // get filter definition from this.state[type]
        var typeFilter = _.find(typeFilters,{by:key})
        
        if (typeof typeFilter !== 'undefined') {
        
          if (typeFilter.type === 'entities') {
            if (_.isArray(filter)) {
              _.each(filter,function(filterItem){
                conditions.push(this.entitiesCondition(typeFilter,filterItem))
              },this)
            } else {
              conditions.push(this.entitiesCondition(typeFilter,filter))              
            }                                             

          } else if(typeFilter.type === 'smart') {
            conditions.push(this.smartCondition(typeFilter,filter))  
            
          } else if(typeFilter.type === 'response') {
            conditions.push(this.responseCondition(typeFilter,filter))  
            
          } else if(typeFilter.type === 'date') {
            if (typeFilter.relation === 'after') {
              conditions.push(this.dateAfterCondition(typeFilter,filter))
            } else if (typeFilter.relation === 'before') {
              conditions.push(this.dateBeforeCondition(typeFilter,filter))
            }

          } else {         
              conditions.push(this.textCondition(typeFilter,filter))
          }
        } else {
          conditions.push(function(entity){      
            return entity[key] === filter
          })
        }
      },this);
      
      
      
      
      
      var entities = conditions.length === 0 
        ? flux.stores[type].entities()
        : flux.stores[type].filter(function(item){        
          return conditions.length === 1 
            ? conditions[0](item)
            : _.reduce(conditions, function(bool, condition){          
              return (typeof bool === 'function' ? bool(item) : bool) && condition(item)
            })
          })
//          <Filters 
//            type = {this.state.types[type]}
//            entities = {entities}
//          />
                
      return (
        <div role='main' >
          <NavTop />              

          <div role='article' className="page-content-issue page-content">
            <Grid className='page-section issue-section-analysis'>
               {this.renderMain()}
            </Grid>
            <Grid className={'page-section page-section-' + type}>
              <EntityTableWrap
                caption = {entities.length + ' Entitites (TODO)'}
                collapsible={false}  
                className={type + '-table'}
                linked = {true}
                cols = {getColumns(type, flux, entities)}            
                rows = {entities} 
                entityType = 'type'
              />
            </Grid>
          </div>
        </div>
      )
    }
  },
  entitiesCondition : function(typeFilter,filterItem){
    var flux = this.getFlux()
    return function(entity){      
      if (typeFilter.relation === 'to') {
        return entity[typeFilter.attr].indexOf(filterItem) > -1
      } else if (typeFilter.relation === 'from') {
        return flux.store(typeFilter.from).filter(function(item){
          return item.id === filterItem && item[typeFilter.attr].indexOf(entity.id) > -1
        }).length > 0                    
      }    
    }
  },
  smartCondition : function(typeFilter,filterItem){
    return function(entity){      
      return (filterItem === 'true' && checkSMART(entity[typeFilter.attr]))
          || (filterItem === 'false' && !checkSMART(entity[typeFilter.attr]))
    }
  },
  responseCondition : function(typeFilter,filterItem){
    return function(entity){      
      return (typeFilter.by === 'accepted' && entity[typeFilter.attr] === 'Accepted')
          || (typeFilter.by === 'rejected' && entity[typeFilter.attr] === 'Rejected')
    }
  },
  textCondition : function(typeFilter,filterItem){
    return function(entity){      
      return entity[typeFilter.attr] === filterItem
    }
  },
  dateAfterCondition : function(typeFilter,filterItem){
    return function(entity){    
      var value = entity[typeFilter.attr]
      return value !== '' 
          && value.format() !== 'Invalid date'
          && value.isAfter(moment(filterItem,'MM-YYYY',true),'month')
    }
  },
  dateBeforeCondition : function(typeFilter,filterItem){
    return function(entity){     
      var value = entity[typeFilter.attr]
      return value !== '' 
          && value.format() !== 'Invalid date'
          && value.isBefore(moment(filterItem,'MM-YYYY',true),'month')
    }
  },
  renderMain : function (){
    var flux = this.getFlux()
    return (
      
        <Row key={0} className="header-row">
          <Col sm={12}>
            <h5 className="header-type">TODO MAIN</h5>
          </Col>
        </Row>
           
    )
  },
  
      


 

})
