// components/explore/agency/agency.js
var React           = require('react')
var Fluxxor         = require('fluxxor')
var StoreWatchMixin = Fluxxor.StoreWatchMixin

//components

// bootstrap
var Row 				= require('react-bootstrap/lib/Row')
var Col 				= require('react-bootstrap/lib/Col')

//helpers

var _               = require('lodash')

var log       = require('debug')('src:components:explore:agency')

module.exports = React.createClass({
  mixins: [
    Fluxxor.FluxMixin(React),
    StoreWatchMixin('agencies', 'actions')
  ],

  getStateFromFlux: function () {
    var flux = this.props.flux
    return {
      entitiesLoading: flux.store('agencies').isLoading(),
      actionsLoading : flux.store("actions").isLoading()      
    }
  },

 render: function () {
    var flux = this.props.flux

    return (
      <Row>
        { this.renderMain() }
      </Row>
    )
  },
  renderMain : function (){
    var flux = this.props.flux
    var entity = flux.stores.agencies.findOne({ id: this.props.params.id })
log('agency',entity)
    return (
      <Col md={12}>
        <p>Agency </p>
        <p>
          { (this.state.entitiesLoading) ? 'loading' : entity.title }
        </p>
      </Col>
    )
  },

 

})
